Memory Leak caused by NullPointerException in ConnectorXAResource.getResourceHandle

I think a null check is missing to avoid this memory leak where TimerTask that check Transaction Timeout aren't removed. Result is an OutOfMemory occured during DB Connection Pool Problems.

The NPE seems to cause a infinite loop in timerthread, we see this stacktraces infinite in the server.log

see stacktrace of the NullPointer + screen of the Memory Leak 
```
[2025-04-03T19:18:46.404541+02:00] [GF 7.0.22] [INFO] [jdbc.exc_table_validation] [jakarta.enterprise.resource.resourceadapter.com.sun.gjc.spi] [tid: _ThreadID=3345 _ThreadName=uf-sdk-26167-3-rm-t-0] [levelValue: 800] [[
  RAR5074 : Table based validation detected invalid connection. Querying the table dual failed. Set resource-adapter log-level to FINE for exception stack trace.]]

[2025-04-03T19:18:46.404783+02:00] [GF 7.0.22] [SEVERE] [poolmgr.system_exception] [jakarta.enterprise.resource.resourceadapter.com.sun.enterprise.resource] [tid: _ThreadID=3345 _ThreadName=uf-sdk-26167-3-rm-t-0] [levelValue: 1000] [[
  RAR5031:System Exception
java.lang.NullPointerException: Cannot invoke "com.sun.enterprise.resource.ResourceHandle.getResourceState()" because "resourceHandle" is null
	at com.sun.enterprise.resource.ConnectorXAResource.getResourceHandle(ConnectorXAResource.java:213)
	at com.sun.enterprise.resource.ConnectorXAResource.rollback(ConnectorXAResource.java:179)
	at com.sun.enterprise.transaction.JavaEETransactionImpl.rollback(JavaEETransactionImpl.java:571)
	at com.sun.enterprise.resource.allocator.LocalTxConnectorAllocator.forceTransactionCompletion(LocalTxConnectorAllocator.java:140)
	at com.sun.enterprise.resource.allocator.LocalTxConnectorAllocator.destroyResource(LocalTxConnectorAllocator.java:110)
	at com.sun.enterprise.resource.pool.ConnectionPool.deleteResource(ConnectionPool.java:1054)
	at com.sun.enterprise.resource.pool.datastructure.RWLockDataStructure.removeAll(RWLockDataStructure.java:305)
	at com.sun.enterprise.resource.pool.ConnectionPool.emptyPool(ConnectionPool.java:1367)
	at com.sun.enterprise.resource.pool.ConnectionPool.doFailAllConnectionsProcessing(ConnectionPool.java:1245)
	at com.sun.enterprise.resource.pool.ConnectionPool.resourceErrorOccurred(ConnectionPool.java:1196)
	at com.sun.enterprise.resource.pool.PoolManagerImpl.putbackResourceToPool(PoolManagerImpl.java:443)
	at com.sun.enterprise.resource.pool.PoolManagerImpl.resourceErrorOccurred(PoolManagerImpl.java:393)
	at com.sun.enterprise.resource.listener.LocalTxConnectionEventListener.connectionErrorOccurred(LocalTxConnectionEventListener.java:101)
	at com.sun.gjc.spi.ManagedConnectionImpl.connectionErrorOccurred(ManagedConnectionImpl.java:785)
	at com.sun.gjc.spi.ManagedConnectionFactoryImpl.getInvalidConnections(ManagedConnectionFactoryImpl.java:300)
	at com.sun.enterprise.resource.allocator.AbstractConnectorAllocator.getInvalidConnections(AbstractConnectorAllocator.java:88)
	at com.sun.enterprise.resource.allocator.AbstractConnectorAllocator.isConnectionValid(AbstractConnectorAllocator.java:99)
	at com.sun.enterprise.resource.pool.ConnectionPool.isConnectionValid(ConnectionPool.java:750)
	at com.sun.enterprise.resource.pool.ConnectionPool.getResourceFromPool(ConnectionPool.java:823)
	at com.sun.enterprise.resource.pool.ConnectionPool.getUnenlistedResource(ConnectionPool.java:716)
	at com.sun.enterprise.resource.pool.ConnectionPool.internalGetResource(ConnectionPool.java:596)
	at com.sun.enterprise.resource.pool.ConnectionPool.getResource(ConnectionPool.java:465)
	at com.sun.enterprise.resource.pool.PoolManagerImpl.getResourceFromPool(PoolManagerImpl.java:243)
	at com.sun.enterprise.resource.pool.PoolManagerImpl.getResource(PoolManagerImpl.java:154)
	at com.sun.enterprise.connectors.ConnectionManagerImpl.getResource(ConnectionManagerImpl.java:355)
	at com.sun.enterprise.connectors.ConnectionManagerImpl.internalGetConnection(ConnectionManagerImpl.java:298)
	at com.sun.enterprise.connectors.ConnectionManagerImpl.allocateConnection(ConnectionManagerImpl.java:183)
	at com.sun.enterprise.connectors.ConnectionManagerImpl.allocateConnection(ConnectionManagerImpl.java:157)
	at com.sun.enterprise.connectors.ConnectionManagerImpl.allocateConnection(ConnectionManagerImpl.java:151)
	at com.sun.gjc.spi.base.AbstractDataSource.getConnection(AbstractDataSource.java:99)
	at org.eclipse.persistence.sessions.JNDIConnector.connect(JNDIConnector.java:140)
	at org.eclipse.persistence.sessions.DatasourceLogin.connectToDatasource(DatasourceLogin.java:174)
	at org.eclipse.persistence.internal.databaseaccess.DatasourceAccessor.connectInternal(DatasourceAccessor.java:375)
	at org.eclipse.persistence.internal.databaseaccess.DatabaseAccessor.connectInternal(DatabaseAccessor.java:318)
	at org.eclipse.persistence.internal.databaseaccess.DatasourceAccessor.reconnect(DatasourceAccessor.java:618)
	at org.eclipse.persistence.internal.databaseaccess.DatabaseAccessor.reconnect(DatabaseAccessor.java:1820)
	at org.eclipse.persistence.internal.databaseaccess.DatasourceAccessor.incrementCallCount(DatasourceAccessor.java:346)
	at org.eclipse.persistence.internal.databaseaccess.DatabaseAccessor.basicExecuteCall(DatabaseAccessor.java:624)
	at org.eclipse.persistence.internal.databaseaccess.DatabaseAccessor.executeCall(DatabaseAccessor.java:569)
	at org.eclipse.persistence.internal.sessions.AbstractSession.basicExecuteCall(AbstractSession.java:2048)
	at org.eclipse.persistence.sessions.server.ClientSession.executeCall(ClientSession.java:311)
	at org.eclipse.persistence.internal.queries.DatasourceCallQueryMechanism.executeCall(DatasourceCallQueryMechanism.java:280)
	at org.eclipse.persistence.internal.queries.DatasourceCallQueryMechanism.executeCall(DatasourceCallQueryMechanism.java:266)
	at org.eclipse.persistence.internal.queries.DatasourceCallQueryMechanism.selectOneRow(DatasourceCallQueryMechanism.java:813)
	at org.eclipse.persistence.internal.queries.ExpressionQueryMechanism.selectOneRowFromTable(ExpressionQueryMechanism.java:2912)
	at org.eclipse.persistence.internal.queries.ExpressionQueryMechanism.selectOneRow(ExpressionQueryMechanism.java:2865)
	at org.eclipse.persistence.queries.ReadObjectQuery.executeObjectLevelReadQuery(ReadObjectQuery.java:563)
	at org.eclipse.persistence.queries.ObjectLevelReadQuery.executeDatabaseQuery(ObjectLevelReadQuery.java:1236)
	at org.eclipse.persistence.queries.DatabaseQuery.execute(DatabaseQuery.java:913)
	at org.eclipse.persistence.queries.ObjectLevelReadQuery.execute(ObjectLevelReadQuery.java:1195)
	at org.eclipse.persistence.queries.ReadObjectQuery.execute(ReadObjectQuery.java:448)
	at org.eclipse.persistence.queries.ObjectLevelReadQuery.executeInUnitOfWork(ObjectLevelReadQuery.java:1283)
	at org.eclipse.persistence.internal.sessions.UnitOfWorkImpl.internalExecuteQuery(UnitOfWorkImpl.java:3031)
	at org.eclipse.persistence.internal.sessions.AbstractSession.executeQuery(AbstractSession.java:1841)
	at org.eclipse.persistence.internal.sessions.AbstractSession.executeQuery(AbstractSession.java:1823)
	at org.eclipse.persistence.internal.sessions.AbstractSession.executeQuery(AbstractSession.java:1773)
	at org.eclipse.persistence.internal.jpa.EntityManagerImpl.executeQuery(EntityManagerImpl.java:999)
	at org.eclipse.persistence.internal.jpa.EntityManagerImpl.findInternal(EntityManagerImpl.java:941)
	at org.eclipse.persistence.internal.jpa.EntityManagerImpl.find(EntityManagerImpl.java:817)
	at com.sun.enterprise.container.common.impl.EntityManagerWrapper.find(EntityManagerWrapper.java:387)
	at uep.game.betting.live.business.LiveManager.betstopAllEvents(LiveManager.java:556)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at org.glassfish.exousia.AuthorizationService.lambda$invokeBeanMethod$1(AuthorizationService.java:506)
	at org.glassfish.exousia.AuthorizationService.runInScope(AuthorizationService.java:586)
	at org.glassfish.exousia.AuthorizationService.invokeBeanMethod(AuthorizationService.java:506)
	at org.glassfish.ejb.security.application.EJBSecurityManager.invoke(EJBSecurityManager.java:297)
	at com.sun.ejb.containers.BaseContainer.invokeBeanMethod(BaseContainer.java:4193)
	at com.sun.ejb.EjbInvocation.invokeBeanMethod(EjbInvocation.java:604)
	at com.sun.ejb.containers.interceptors.AroundInvokeChainImpl.invokeNext(InterceptorManager.java:706)
	at com.sun.ejb.EjbInvocation.proceed(EjbInvocation.java:553)
	at org.jboss.weld.module.ejb.AbstractEJBRequestScopeActivationInterceptor.aroundInvoke(AbstractEJBRequestScopeActivationInterceptor.java:87)
	at org.jboss.weld.module.ejb.SessionBeanInterceptor.aroundInvoke(SessionBeanInterceptor.java:52)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at com.sun.ejb.containers.interceptors.AroundInvokeInterceptor.intercept(InterceptorManager.java:766)
	at com.sun.ejb.containers.interceptors.AroundInvokeChainImpl.invokeNext(InterceptorManager.java:706)
	at com.sun.ejb.EjbInvocation.proceed(EjbInvocation.java:553)
	at com.sun.ejb.containers.interceptors.SystemInterceptorProxy.doCall(SystemInterceptorProxy.java:122)
	at com.sun.ejb.containers.interceptors.SystemInterceptorProxy.aroundInvoke(SystemInterceptorProxy.java:99)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at com.sun.ejb.containers.interceptors.AroundInvokeInterceptor.intercept(InterceptorManager.java:766)
	at com.sun.ejb.containers.interceptors.AroundInvokeChainImpl.invokeNext(InterceptorManager.java:706)
	at com.sun.ejb.containers.interceptors.InterceptorManager.intercept(InterceptorManager.java:307)
	at com.sun.ejb.containers.BaseContainer.__intercept(BaseContainer.java:4168)
	at com.sun.ejb.containers.BaseContainer.intercept(BaseContainer.java:4158)
	at com.sun.ejb.containers.EJBLocalObjectInvocationHandler.invoke(EJBLocalObjectInvocationHandler.java:169)
	at com.sun.ejb.containers.EJBLocalObjectInvocationHandlerDelegate.invoke(EJBLocalObjectInvocationHandlerDelegate.java:61)
	at jdk.proxy47/jdk.proxy47.$Proxy564.betstopAllEvents(Unknown Source)
	at uep.game.betting.live.business.__EJB31_Generated__LiveManager__Intf____Bean__.betstopAllEvents(Unknown Source)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at org.jboss.weld.util.reflection.Reflections.invokeAndUnwrap(Reflections.java:412)
	at org.jboss.weld.module.ejb.EnterpriseBeanProxyMethodHandler.invoke(EnterpriseBeanProxyMethodHandler.java:133)
	at org.jboss.weld.bean.proxy.EnterpriseTargetBeanInstance.invoke(EnterpriseTargetBeanInstance.java:54)
	at org.jboss.weld.module.ejb.InjectionPointPropagatingEnterpriseTargetBeanInstance.invoke(InjectionPointPropagatingEnterpriseTargetBeanInstance.java:69)
	at org.jboss.weld.bean.proxy.ProxyMethodHandler.invoke(ProxyMethodHandler.java:109)
	at uep.game.betting.live.business.LiveManager$Proxy$_$$_Weld$EnterpriseProxy$.betstopAllEvents(Unknown Source)
	at uep.game.betting.live.business.AbstractExternalFeedHandler.executeBetstop(AbstractExternalFeedHandler.java:341)
	at uep.game.betting.live.business.UnifiedFeedGlobalEventsListener.onProducerStatusChange(UnifiedFeedGlobalEventsListener.java:45)
	at com.sportradar.unifiedodds.sdk.impl.recovery.RecoveryManagerImpl.handleProducerStateChange(RecoveryManagerImpl.java:1059)
	at com.sportradar.unifiedodds.sdk.impl.recovery.RecoveryManagerImpl.flagProducerDown(RecoveryManagerImpl.java:956)
	at com.sportradar.unifiedodds.sdk.impl.recovery.RecoveryManagerImpl.onTimerElapsed(RecoveryManagerImpl.java:496)
	at com.sportradar.unifiedodds.sdk.impl.recovery.SingleRecoveryManagerSupervisor.lambda$scheduleSupervisionJob$0(SingleRecoveryManagerSupervisor.java:73)
	at com.sportradar.unifiedodds.sdk.impl.util.MdcScheduledExecutorService.lambda$setMdcContext$1(MdcScheduledExecutorService.java:499)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:572)
	at java.base/java.util.concurrent.FutureTask.runAndReset(FutureTask.java:358)
	at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:305)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
	at java.base/java.lang.Thread.run(Thread.java:1583)
]]
```
![Image](https://github.com/user-attachments/assets/5e03d03b-def1-4370-941d-21de2f958ac2)

we haven't used prevous version of glassfish so I don't know if this is new in 7.0.22

Exceptions before were. I think the cause is that the database was not responding see other stacktrace below 

[2025-04-02T18:30:07.757096+02:00] [GF 7.0.22] [WARNING] [AS-EJB-00056] [jakarta.enterprise.ejb.container] [tid: _ThreadID=877 _ThreadName=liveBettingPool-2-thread-3] [levelValue: 900] [[
  A system exception occurred during an invocation on EJB SportEventManager, method: public void uep.game.betting.dynamic.business.SportEventManager.releaseEventWhenLive(uep.game.betting.dynamic.persistence.Event) throws uep.core.interfaces.api.ApplicationException]]

[2025-04-02T18:30:07.757228+02:00] [GF 7.0.22] [WARNING] [] [jakarta.enterprise.ejb.container] [tid: _ThreadID=877 _ThreadName=liveBettingPool-2-thread-3] [levelValue: 900] [[
  jakarta.ejb.TransactionRolledbackLocalException: Exception thrown from bean
	at com.sun.ejb.containers.EJBContainerTransactionManager.checkExceptionClientTx(EJBContainerTransactionManager.java:686)
	at com.sun.ejb.containers.EJBContainerTransactionManager.postInvokeTx(EJBContainerTransactionManager.java:528)
	at com.sun.ejb.containers.BaseContainer.postInvokeTx(BaseContainer.java:3988)
	at com.sun.ejb.containers.BaseContainer.postInvoke(BaseContainer.java:1807)
	at com.sun.ejb.containers.BaseContainer.postInvoke(BaseContainer.java:1780)
	at com.sun.ejb.containers.EJBLocalObjectInvocationHandler.invoke(EJBLocalObjectInvocationHandler.java:176)
	at com.sun.ejb.containers.EJBLocalObjectInvocationHandlerDelegate.invoke(EJBLocalObjectInvocationHandlerDelegate.java:61)
	at jdk.proxy47/jdk.proxy47.$Proxy596.releaseEventWhenLive(Unknown Source)
	at uep.game.betting.dynamic.business.__EJB31_Generated__SportEventManager__Intf____Bean__.releaseEventWhenLive(Unknown Source)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at org.jboss.weld.util.reflection.Reflections.invokeAndUnwrap(Reflections.java:412)
	at org.jboss.weld.module.ejb.EnterpriseBeanProxyMethodHandler.invoke(EnterpriseBeanProxyMethodHandler.java:133)
	at org.jboss.weld.bean.proxy.EnterpriseTargetBeanInstance.invoke(EnterpriseTargetBeanInstance.java:54)
	at org.jboss.weld.module.ejb.InjectionPointPropagatingEnterpriseTargetBeanInstance.invoke(InjectionPointPropagatingEnterpriseTargetBeanInstance.java:69)
	at org.jboss.weld.bean.proxy.ProxyMethodHandler.invoke(ProxyMethodHandler.java:109)
	at uep.game.betting.dynamic.business.SportEventManager$Proxy$_$$_Weld$EnterpriseProxy$.releaseEventWhenLive(Unknown Source)
	at uep.game.betting.live.business.UnifiedFeedTakeoverManager.processMessage(UnifiedFeedTakeoverManager.java:215)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at org.glassfish.exousia.AuthorizationService.lambda$invokeBeanMethod$1(AuthorizationService.java:506)
	at org.glassfish.exousia.AuthorizationService.runInScope(AuthorizationService.java:586)
	at org.glassfish.exousia.AuthorizationService.invokeBeanMethod(AuthorizationService.java:506)
	at org.glassfish.ejb.security.application.EJBSecurityManager.invoke(EJBSecurityManager.java:297)
	at com.sun.ejb.containers.BaseContainer.invokeBeanMethod(BaseContainer.java:4193)
	at com.sun.ejb.EjbInvocation.invokeBeanMethod(EjbInvocation.java:604)
	at com.sun.ejb.containers.interceptors.AroundInvokeChainImpl.invokeNext(InterceptorManager.java:706)
	at com.sun.ejb.EjbInvocation.proceed(EjbInvocation.java:553)
	at org.jboss.weld.module.ejb.AbstractEJBRequestScopeActivationInterceptor.aroundInvoke(AbstractEJBRequestScopeActivationInterceptor.java:78)
	at org.jboss.weld.module.ejb.SessionBeanInterceptor.aroundInvoke(SessionBeanInterceptor.java:52)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at com.sun.ejb.containers.interceptors.AroundInvokeInterceptor.intercept(InterceptorManager.java:766)
	at com.sun.ejb.containers.interceptors.AroundInvokeChainImpl.invokeNext(InterceptorManager.java:706)
	at com.sun.ejb.EjbInvocation.proceed(EjbInvocation.java:553)
	at com.sun.ejb.containers.interceptors.SystemInterceptorProxy.doCall(SystemInterceptorProxy.java:122)
	at com.sun.ejb.containers.interceptors.SystemInterceptorProxy.aroundInvoke(SystemInterceptorProxy.java:99)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at com.sun.ejb.containers.interceptors.AroundInvokeInterceptor.intercept(InterceptorManager.java:766)
	at com.sun.ejb.containers.interceptors.AroundInvokeChainImpl.invokeNext(InterceptorManager.java:706)
	at com.sun.ejb.containers.interceptors.InterceptorManager.intercept(InterceptorManager.java:307)
	at com.sun.ejb.containers.BaseContainer.__intercept(BaseContainer.java:4168)
	at com.sun.ejb.containers.BaseContainer.intercept(BaseContainer.java:4158)
	at com.sun.ejb.containers.EJBLocalObjectInvocationHandler.invoke(EJBLocalObjectInvocationHandler.java:169)
	at com.sun.ejb.containers.EJBLocalObjectInvocationHandlerDelegate.invoke(EJBLocalObjectInvocationHandlerDelegate.java:61)
	at jdk.proxy47/jdk.proxy47.$Proxy560.processMessage(Unknown Source)
	at uep.game.betting.live.business.__EJB31_Generated__UnifiedFeedTakeoverManager__Intf____Bean__.processMessage(Unknown Source)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at org.jboss.weld.util.reflection.Reflections.invokeAndUnwrap(Reflections.java:412)
	at org.jboss.weld.module.ejb.EnterpriseBeanProxyMethodHandler.invoke(EnterpriseBeanProxyMethodHandler.java:133)
	at org.jboss.weld.bean.proxy.EnterpriseTargetBeanInstance.invoke(EnterpriseTargetBeanInstance.java:54)
	at org.jboss.weld.module.ejb.InjectionPointPropagatingEnterpriseTargetBeanInstance.invoke(InjectionPointPropagatingEnterpriseTargetBeanInstance.java:69)
	at org.jboss.weld.bean.proxy.ProxyMethodHandler.invoke(ProxyMethodHandler.java:109)
	at uep.game.betting.live.business.UnifiedFeedTakeoverManager$Proxy$_$$_Weld$EnterpriseProxy$.processMessage(Unknown Source)
	at uep.game.betting.live.business.LiveManager.processMessage(LiveManager.java:416)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at org.glassfish.exousia.AuthorizationService.lambda$invokeBeanMethod$1(AuthorizationService.java:506)
	at org.glassfish.exousia.AuthorizationService.runInScope(AuthorizationService.java:586)
	at org.glassfish.exousia.AuthorizationService.invokeBeanMethod(AuthorizationService.java:506)
	at org.glassfish.ejb.security.application.EJBSecurityManager.invoke(EJBSecurityManager.java:297)
	at com.sun.ejb.containers.BaseContainer.invokeBeanMethod(BaseContainer.java:4193)
	at com.sun.ejb.EjbInvocation.invokeBeanMethod(EjbInvocation.java:604)
	at com.sun.ejb.containers.interceptors.AroundInvokeChainImpl.invokeNext(InterceptorManager.java:706)
	at com.sun.ejb.EjbInvocation.proceed(EjbInvocation.java:553)
	at org.jboss.weld.module.ejb.AbstractEJBRequestScopeActivationInterceptor.aroundInvoke(AbstractEJBRequestScopeActivationInterceptor.java:87)
	at org.jboss.weld.module.ejb.SessionBeanInterceptor.aroundInvoke(SessionBeanInterceptor.java:52)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at com.sun.ejb.containers.interceptors.AroundInvokeInterceptor.intercept(InterceptorManager.java:766)
	at com.sun.ejb.containers.interceptors.AroundInvokeChainImpl.invokeNext(InterceptorManager.java:706)
	at com.sun.ejb.EjbInvocation.proceed(EjbInvocation.java:553)
	at com.sun.ejb.containers.interceptors.SystemInterceptorProxy.doCall(SystemInterceptorProxy.java:122)
	at com.sun.ejb.containers.interceptors.SystemInterceptorProxy.aroundInvoke(SystemInterceptorProxy.java:99)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at com.sun.ejb.containers.interceptors.AroundInvokeInterceptor.intercept(InterceptorManager.java:766)
	at com.sun.ejb.containers.interceptors.AroundInvokeChainImpl.invokeNext(InterceptorManager.java:706)
	at com.sun.ejb.containers.interceptors.InterceptorManager.intercept(InterceptorManager.java:307)
	at com.sun.ejb.containers.BaseContainer.__intercept(BaseContainer.java:4168)
	at com.sun.ejb.containers.BaseContainer.intercept(BaseContainer.java:4158)
	at com.sun.ejb.containers.EJBLocalObjectInvocationHandler.invoke(EJBLocalObjectInvocationHandler.java:169)
	at com.sun.ejb.containers.EJBLocalObjectInvocationHandlerDelegate.invoke(EJBLocalObjectInvocationHandlerDelegate.java:61)
	at jdk.proxy47/jdk.proxy47.$Proxy564.processMessage(Unknown Source)
	at uep.game.betting.live.business.__EJB31_Generated__LiveManager__Intf____Bean__.processMessage(Unknown Source)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at org.jboss.weld.util.reflection.Reflections.invokeAndUnwrap(Reflections.java:412)
	at org.jboss.weld.module.ejb.EnterpriseBeanProxyMethodHandler.invoke(EnterpriseBeanProxyMethodHandler.java:133)
	at org.jboss.weld.bean.proxy.EnterpriseTargetBeanInstance.invoke(EnterpriseTargetBeanInstance.java:54)
	at org.jboss.weld.module.ejb.InjectionPointPropagatingEnterpriseTargetBeanInstance.invoke(InjectionPointPropagatingEnterpriseTargetBeanInstance.java:69)
	at org.jboss.weld.bean.proxy.ProxyMethodHandler.invoke(ProxyMethodHandler.java:109)
	at uep.game.betting.live.business.LiveManager$Proxy$_$$_Weld$EnterpriseProxy$.processMessage(Unknown Source)
	at uep.game.betting.live.business.LiveBettingWork.tryComputeMessage(LiveBettingWork.java:182)
	at uep.game.betting.live.business.LiveBettingWork.run(LiveBettingWork.java:119)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:572)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:317)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
	at java.base/java.lang.Thread.run(Thread.java:1583)
Caused by: jakarta.ejb.EJBTransactionRolledbackException
	at com.sun.ejb.containers.BaseContainer.mapLocal3xException(BaseContainer.java:2048)
	at com.sun.ejb.containers.BaseContainer.postInvoke(BaseContainer.java:1858)
	at com.sun.ejb.containers.BaseContainer.postInvoke(BaseContainer.java:1780)
	at com.sun.ejb.containers.EJBLocalObjectInvocationHandler.invoke(EJBLocalObjectInvocationHandler.java:176)
	at com.sun.ejb.containers.EJBLocalObjectInvocationHandlerDelegate.invoke(EJBLocalObjectInvocationHandlerDelegate.java:61)
	at jdk.proxy47/jdk.proxy47.$Proxy583.sendAutomaticReleaseProblemMail(Unknown Source)
	at uep.game.betting.dynamic.business.__EJB31_Generated__SportEventTxManager__Intf____Bean__.sendAutomaticReleaseProblemMail(Unknown Source)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at org.jboss.weld.util.reflection.Reflections.invokeAndUnwrap(Reflections.java:412)
	at org.jboss.weld.module.ejb.EnterpriseBeanProxyMethodHandler.invoke(EnterpriseBeanProxyMethodHandler.java:133)
	at org.jboss.weld.bean.proxy.EnterpriseTargetBeanInstance.invoke(EnterpriseTargetBeanInstance.java:54)
	at org.jboss.weld.module.ejb.InjectionPointPropagatingEnterpriseTargetBeanInstance.invoke(InjectionPointPropagatingEnterpriseTargetBeanInstance.java:69)
	at org.jboss.weld.bean.proxy.ProxyMethodHandler.invoke(ProxyMethodHandler.java:109)
	at uep.game.betting.dynamic.business.SportEventTxManager$Proxy$_$$_Weld$EnterpriseProxy$.sendAutomaticReleaseProblemMail(Unknown Source)
	at uep.game.betting.dynamic.business.SportEventManager.releaseEvents(SportEventManager.java:636)
	at uep.game.betting.dynamic.business.SportEventManager.releaseEventWhenLive(SportEventManager.java:739)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at org.glassfish.exousia.AuthorizationService.lambda$invokeBeanMethod$1(AuthorizationService.java:506)
	at org.glassfish.exousia.AuthorizationService.runInScope(AuthorizationService.java:586)
	at org.glassfish.exousia.AuthorizationService.invokeBeanMethod(AuthorizationService.java:506)
	at org.glassfish.ejb.security.application.EJBSecurityManager.invoke(EJBSecurityManager.java:297)
	at com.sun.ejb.containers.BaseContainer.invokeBeanMethod(BaseContainer.java:4193)
	at com.sun.ejb.EjbInvocation.invokeBeanMethod(EjbInvocation.java:604)
	at com.sun.ejb.containers.interceptors.AroundInvokeChainImpl.invokeNext(InterceptorManager.java:706)
	at com.sun.ejb.EjbInvocation.proceed(EjbInvocation.java:553)
	at org.jboss.weld.module.ejb.AbstractEJBRequestScopeActivationInterceptor.aroundInvoke(AbstractEJBRequestScopeActivationInterceptor.java:78)
	at org.jboss.weld.module.ejb.SessionBeanInterceptor.aroundInvoke(SessionBeanInterceptor.java:52)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at com.sun.ejb.containers.interceptors.AroundInvokeInterceptor.intercept(InterceptorManager.java:766)
	at com.sun.ejb.containers.interceptors.AroundInvokeChainImpl.invokeNext(InterceptorManager.java:706)
	at com.sun.ejb.EjbInvocation.proceed(EjbInvocation.java:553)
	at com.sun.ejb.containers.interceptors.SystemInterceptorProxy.doCall(SystemInterceptorProxy.java:122)
	at com.sun.ejb.containers.interceptors.SystemInterceptorProxy.aroundInvoke(SystemInterceptorProxy.java:99)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at com.sun.ejb.containers.interceptors.AroundInvokeInterceptor.intercept(InterceptorManager.java:766)
	at com.sun.ejb.containers.interceptors.AroundInvokeChainImpl.invokeNext(InterceptorManager.java:706)
	at com.sun.ejb.containers.interceptors.InterceptorManager.intercept(InterceptorManager.java:307)
	at com.sun.ejb.containers.BaseContainer.__intercept(BaseContainer.java:4168)
	at com.sun.ejb.containers.BaseContainer.intercept(BaseContainer.java:4158)
	at com.sun.ejb.containers.EJBLocalObjectInvocationHandler.invoke(EJBLocalObjectInvocationHandler.java:169)
	... 96 more
Caused by: jakarta.ejb.TransactionRolledbackLocalException: Client's transaction aborted
	at com.sun.ejb.containers.EJBContainerTransactionManager.useClientTx(EJBContainerTransactionManager.java:368)
	at com.sun.ejb.containers.EJBContainerTransactionManager.preInvokeTx(EJBContainerTransactionManager.java:248)
	at com.sun.ejb.containers.BaseContainer.preInvokeTx(BaseContainer.java:3952)
	at com.sun.ejb.containers.BaseContainer.preInvoke(BaseContainer.java:1723)
	at com.sun.ejb.containers.EJBLocalObjectInvocationHandler.invoke(EJBLocalObjectInvocationHandler.java:168)
	... 136 more
]]





[2025-04-03T05:10:05.108000+02:00] [GF 7.0.22] [INFO] [] [oracle.jdbc] [tid: _ThreadID=230 _ThreadName=__ejb-thread-pool1] [levelValue: 800] [[
  entering args (oracle.jdbc.datasource.impl.OracleDataSource$1@20ecab8f)]]

[2025-04-03T05:10:05.109000+02:00] [GF 7.0.22] [INFO] [] [oracle.jdbc] [tid: _ThreadID=230 _ThreadName=__ejb-thread-pool1] [levelValue: 800] [[
  traceId=7E9CF684. ]]

[2025-04-03T05:10:05.110000+02:00] [GF 7.0.22] [INFO] [] [oracle.jdbc] [tid: _ThreadID=230 _ThreadName=__ejb-thread-pool1] [levelValue: 800] [[
  Session Attributes: 
sdu=8192, tdu=2097152
nt: host=rwdbuep2tfm1.lottery.co.at, port=1522, socketOptions={0=YES, 1=NO, 2=20000, 38=TLS, 40=false, 8=/server/glassfish7/glassfish/domains/uepdomain/config/keystore.jks, 10=oracle.jdbc.internal.OpaqueString@5560f4db, 11=/server/glassfish7/glassfish/domains/uepdomain/config/cacerts.jks, 13=oracle.jdbc.internal.OpaqueString@5560f4db, 17=0, 18=false, 20=true, 23=40, 24=50, 25=0}
    socket=null
client profile={oracle.net.encryption_types_client=(), oracle.net.crypto_seed=, oracle.net.authentication_services=(), oracle.net.setFIPSMode=false, oracle.net.kerberos5_mutual_authentication=false, oracle.net.encryption_client=ACCEPTED, oracle.net.crypto_checksum_client=ACCEPTED, oracle.net.crypto_checksum_types_client=()}
connection options=[host=rwdbuep2tfm1.lottery.co.at port=1522 protocol=TCP service_name=UEP2TFM1 addr=(ADDRESS=(PROTOCOL=TCP)(HOST=rwdbuep2tfm1.lottery.co.at)(PORT=1522)) conn_data=(DESCRIPTION=(ADDRESS=(PROTOCOL=TCP)(PORT=1522)(HOST=192.168.53.68)(HOSTNAME=rwdbuep2tfm1.lottery.co.at))(CONNECT_DATA=(CID=(PROGRAM=GlassFishMain)(HOST=rwueptst7.lottery.co.at)(USER=bea))(SERVICE_NAME=UEP2TFM1))) done=true]
onBreakReset=false, dataEOF=false, negotiatedOptions=0xc41, connected=false
TTIINIT enabled=false, TTC cookie enabled=false
]]

[2025-04-03T05:10:05.111000+02:00] [GF 7.0.22] [INFO] [] [oracle.jdbc] [tid: _ThreadID=230 _ThreadName=__ejb-thread-pool1] [levelValue: 800] [[
  traceId=7E9CF684, anoEnabled=true. ]]

[2025-04-03T05:10:05.114000+02:00] [GF 7.0.22] [INFO] [] [oracle.jdbc] [tid: _ThreadID=230 _ThreadName=__ejb-thread-pool1] [levelValue: 800] [[
  Got Resend, SessionTraceId = 7E9CF684]]

[2025-04-03T05:10:05.117000+02:00] [GF 7.0.22] [INFO] [] [oracle.jdbc] [tid: _ThreadID=230 _ThreadName=__ejb-thread-pool1] [levelValue: 800] [[
  Connection established. Cleared conn strategy stack]]

[2025-04-03T05:10:05.119000+02:00] [GF 7.0.22] [INFO] [] [oracle.jdbc] [tid: _ThreadID=230 _ThreadName=__ejb-thread-pool1] [levelValue: 800] [[
  Sending break marker, SessionTraceId = 7E9CF684]]

[2025-04-03T05:10:05.130000+02:00] [GF 7.0.22] [INFO] [] [oracle.jdbc] [tid: _ThreadID=230 _ThreadName=__ejb-thread-pool1] [levelValue: 800] [[
  returning void]]

[2025-04-03T05:15:07.073000+02:00] [GF 7.0.22] [INFO] [] [oracle.jdbc] [tid: _ThreadID=215 _ThreadName=__ejb-thread-pool1] [levelValue: 800] [[
  SO_TIMEOUT interrupt timer cancelled null]]

[2025-04-03T05:15:07.074000+02:00] [GF 7.0.22] [INFO] [] [oracle.jdbc] [tid: _ThreadID=215 _ThreadName=__ejb-thread-pool1] [levelValue: 800] [[
  Sending urgent marker.  SessionTraceId=7E9CF684. ]]

[2025-04-03T05:20:06.419000+02:00] [GF 7.0.22] [INFO] [] [oracle.jdbc] [tid: _ThreadID=224 _ThreadName=__ejb-thread-pool1] [levelValue: 800] [[
  java.sql.SQLRecoverableException: ORA-03113: database connection closed by peer (connection_type=client, source=/192.168.56.208:45924, target=rwdbuep2tfm1.lottery.co.at:1522, protocol=TCP, service=UEP2TFM1, detecting_end=client, server_type=unknown, client_program=GlassFishMain, server_process=29284, session=790.28240, user=UEP2_FK7, last_rpc=Execute query, connection_id=qjR4TE3tStqL3IXLUK3icA==, nne_encryption=null, nne_checksumming=null, authentication=O5LOGON)
https://docs.oracle.com/error-help/db/ora-03113/
	at oracle.jdbc.driver.T4CPreparedStatement.executeForDescribe(T4CPreparedStatement.java:1246)
	at oracle.jdbc.driver.OracleStatement.prepareDefineBufferAndExecute(OracleStatement.java:1413)
	at oracle.jdbc.driver.OracleStatement.executeMaybeDescribe(OracleStatement.java:1287)
	at oracle.jdbc.driver.OracleStatement.executeSQLSelect(OracleStatement.java:1844)
	at oracle.jdbc.driver.OracleStatement.doExecuteWithTimeout(OracleStatement.java:1620)
	at oracle.jdbc.driver.OraclePreparedStatement.executeInternal(OraclePreparedStatement.java:3961)
	at oracle.jdbc.driver.OraclePreparedStatement.executeQuery(OraclePreparedStatement.java:4148)
	at oracle.jdbc.driver.OraclePreparedStatementWrapper.executeQuery(OraclePreparedStatementWrapper.java:1117)
	at com.sun.gjc.spi.jdbc40.PreparedStatementWrapper40.executeQuery(PreparedStatementWrapper40.java:656)
	at org.eclipse.persistence.internal.databaseaccess.DatabaseAccessor.executeSelect(DatabaseAccessor.java:1026)
	at org.eclipse.persistence.internal.databaseaccess.DatabaseAccessor.basicExecuteCall(DatabaseAccessor.java:662)
	at org.eclipse.persistence.internal.databaseaccess.DatabaseAccessor.executeCall(DatabaseAccessor.java:569)
	at org.eclipse.persistence.internal.sessions.AbstractSession.basicExecuteCall(AbstractSession.java:2048)
	at org.eclipse.persistence.sessions.server.ServerSession.executeCall(ServerSession.java:611)
	at org.eclipse.persistence.sessions.server.ClientSession.executeCall(ClientSession.java:263)
	at org.eclipse.persistence.internal.queries.DatasourceCallQueryMechanism.executeCall(DatasourceCallQueryMechanism.java:280)
	at org.eclipse.persistence.internal.queries.DatasourceCallQueryMechanism.executeCall(DatasourceCallQueryMechanism.java:266)
	at org.eclipse.persistence.internal.queries.DatasourceCallQueryMechanism.executeSelectCall(DatasourceCallQueryMechanism.java:352)
	at org.eclipse.persistence.internal.queries.DatasourceCallQueryMechanism.selectAllRows(DatasourceCallQueryMechanism.java:792)
	at org.eclipse.persistence.internal.queries.ExpressionQueryMechanism.selectAllRowsFromTable(ExpressionQueryMechanism.java:2848)
	at org.eclipse.persistence.internal.queries.ExpressionQueryMechanism.selectAllRows(ExpressionQueryMechanism.java:2801)
	at org.eclipse.persistence.queries.ReadAllQuery.executeObjectLevelReadQuery(ReadAllQuery.java:584)
	at org.eclipse.persistence.queries.ObjectLevelReadQuery.executeDatabaseQuery(ObjectLevelReadQuery.java:1236)
	at org.eclipse.persistence.queries.DatabaseQuery.execute(DatabaseQuery.java:913)
	at org.eclipse.persistence.queries.ObjectLevelReadQuery.execute(ObjectLevelReadQuery.java:1195)
	at org.eclipse.persistence.queries.ReadAllQuery.execute(ReadAllQuery.java:485)
	at org.eclipse.persistence.queries.ObjectLevelReadQuery.executeInUnitOfWork(ObjectLevelReadQuery.java:1283)
	at org.eclipse.persistence.internal.sessions.UnitOfWorkImpl.internalExecuteQuery(UnitOfWorkImpl.java:3031)
	at org.eclipse.persistence.internal.sessions.AbstractSession.executeQuery(AbstractSession.java:1841)
	at org.eclipse.persistence.internal.sessions.AbstractSession.executeQuery(AbstractSession.java:1823)
	at org.eclipse.persistence.internal.indirection.NoIndirectionPolicy.valueFromQuery(NoIndirectionPolicy.java:352)
	at org.eclipse.persistence.mappings.ForeignReferenceMapping.valueFromRowInternal(ForeignReferenceMapping.java:2343)
	at org.eclipse.persistence.mappings.ForeignReferenceMapping.valueFromRow(ForeignReferenceMapping.java:2184)
	at org.eclipse.persistence.mappings.ForeignReferenceMapping.buildCloneFromRow(ForeignReferenceMapping.java:341)
	at org.eclipse.persistence.internal.descriptors.ObjectBuilder.buildAttributesIntoWorkingCopyClone(ObjectBuilder.java:2111)
	at org.eclipse.persistence.internal.descriptors.ObjectBuilder.buildWorkingCopyCloneFromRow(ObjectBuilder.java:2364)
	at org.eclipse.persistence.internal.descriptors.ObjectBuilder.buildObjectInUnitOfWork(ObjectBuilder.java:958)
	at org.eclipse.persistence.internal.descriptors.ObjectBuilder.buildObjectInternal(ObjectBuilder.java:844)
	at org.eclipse.persistence.internal.descriptors.ObjectBuilder.buildObject(ObjectBuilder.java:799)
	at org.eclipse.persistence.internal.descriptors.ObjectBuilder.buildObject(ObjectBuilder.java:777)
	at org.eclipse.persistence.queries.ObjectLevelReadQuery.buildObject(ObjectLevelReadQuery.java:863)
	at org.eclipse.persistence.queries.ReadObjectQuery.registerResultInUnitOfWork(ReadObjectQuery.java:896)
	at org.eclipse.persistence.queries.ReadObjectQuery.executeObjectLevelReadQuery(ReadObjectQuery.java:570)
	at org.eclipse.persistence.queries.ObjectLevelReadQuery.executeDatabaseQuery(ObjectLevelReadQuery.java:1236)
	at org.eclipse.persistence.queries.DatabaseQuery.execute(DatabaseQuery.java:913)
	at org.eclipse.persistence.queries.ObjectLevelReadQuery.execute(ObjectLevelReadQuery.java:1195)
	at org.eclipse.persistence.queries.ReadObjectQuery.execute(ReadObjectQuery.java:448)
	at org.eclipse.persistence.queries.ObjectLevelReadQuery.executeInUnitOfWork(ObjectLevelReadQuery.java:1283)
	at org.eclipse.persistence.internal.sessions.UnitOfWorkImpl.internalExecuteQuery(UnitOfWorkImpl.java:3031)
	at org.eclipse.persistence.internal.sessions.AbstractSession.executeQuery(AbstractSession.java:1841)
	at org.eclipse.persistence.internal.sessions.AbstractSession.executeQuery(AbstractSession.java:1823)
	at org.eclipse.persistence.internal.indirection.NoIndirectionPolicy.valueFromQuery(NoIndirectionPolicy.java:352)
	at org.eclipse.persistence.mappings.ForeignReferenceMapping.valueFromRowInternal(ForeignReferenceMapping.java:2343)
	at org.eclipse.persistence.mappings.OneToOneMapping.valueFromRowInternal(OneToOneMapping.java:1874)
	at org.eclipse.persistence.mappings.ForeignReferenceMapping.valueFromRow(ForeignReferenceMapping.java:2184)
	at org.eclipse.persistence.mappings.ForeignReferenceMapping.buildCloneFromRow(ForeignReferenceMapping.java:341)
	at org.eclipse.persistence.internal.descriptors.ObjectBuilder.buildAttributesIntoWorkingCopyClone(ObjectBuilder.java:2111)
	at org.eclipse.persistence.internal.descriptors.ObjectBuilder.buildWorkingCopyCloneFromRow(ObjectBuilder.java:2364)
	at org.eclipse.persistence.internal.descriptors.ObjectBuilder.buildObjectInUnitOfWork(ObjectBuilder.java:958)
	at org.eclipse.persistence.internal.descriptors.ObjectBuilder.buildObjectInternal(ObjectBuilder.java:844)
	at org.eclipse.persistence.internal.descriptors.ObjectBuilder.buildObject(ObjectBuilder.java:799)
	at org.eclipse.persistence.internal.descriptors.ObjectBuilder.buildObject(ObjectBuilder.java:777)
	at org.eclipse.persistence.queries.ObjectLevelReadQuery.buildObject(ObjectLevelReadQuery.java:863)
	at org.eclipse.persistence.queries.ReadAllQuery.registerResultInUnitOfWork(ReadAllQuery.java:987)
	at org.eclipse.persistence.queries.ReadAllQuery.executeObjectLevelReadQuery(ReadAllQuery.java:598)
	at org.eclipse.persistence.queries.ObjectLevelReadQuery.executeDatabaseQuery(ObjectLevelReadQuery.java:1236)
	at org.eclipse.persistence.queries.DatabaseQuery.execute(DatabaseQuery.java:913)
	at org.eclipse.persistence.queries.ObjectLevelReadQuery.execute(ObjectLevelReadQuery.java:1195)
	at org.eclipse.persistence.queries.ReadAllQuery.execute(ReadAllQuery.java:485)
	at org.eclipse.persistence.queries.ObjectLevelReadQuery.executeInUnitOfWork(ObjectLevelReadQuery.java:1283)
	at org.eclipse.persistence.internal.sessions.UnitOfWorkImpl.internalExecuteQuery(UnitOfWorkImpl.java:3031)
	at org.eclipse.persistence.internal.sessions.AbstractSession.executeQuery(AbstractSession.java:1841)
	at org.eclipse.persistence.internal.sessions.AbstractSession.executeQuery(AbstractSession.java:1823)
	at org.eclipse.persistence.internal.indirection.NoIndirectionPolicy.valueFromQuery(NoIndirectionPolicy.java:352)
	at org.eclipse.persistence.mappings.ForeignReferenceMapping.valueFromRowInternal(ForeignReferenceMapping.java:2343)
	at org.eclipse.persistence.mappings.ForeignReferenceMapping.valueFromRow(ForeignReferenceMapping.java:2184)
	at org.eclipse.persistence.mappings.ForeignReferenceMapping.buildCloneFromRow(ForeignReferenceMapping.java:341)
	at org.eclipse.persistence.internal.descriptors.ObjectBuilder.buildAttributesIntoWorkingCopyClone(ObjectBuilder.java:2111)
	at org.eclipse.persistence.internal.descriptors.ObjectBuilder.buildWorkingCopyCloneFromRow(ObjectBuilder.java:2364)
	at org.eclipse.persistence.internal.descriptors.ObjectBuilder.buildObjectInUnitOfWork(ObjectBuilder.java:958)
	at org.eclipse.persistence.internal.descriptors.ObjectBuilder.buildObjectInternal(ObjectBuilder.java:844)
	at org.eclipse.persistence.internal.descriptors.ObjectBuilder.buildObject(ObjectBuilder.java:799)
	at org.eclipse.persistence.internal.descriptors.ObjectBuilder.buildObject(ObjectBuilder.java:777)
	at org.eclipse.persistence.queries.ObjectLevelReadQuery.buildObject(ObjectLevelReadQuery.java:863)
	at org.eclipse.persistence.queries.ReadAllQuery.registerResultInUnitOfWork(ReadAllQuery.java:987)
	at org.eclipse.persistence.queries.ReadAllQuery.executeObjectLevelReadQuery(ReadAllQuery.java:598)
	at org.eclipse.persistence.queries.ObjectLevelReadQuery.executeDatabaseQuery(ObjectLevelReadQuery.java:1236)
	at org.eclipse.persistence.queries.DatabaseQuery.execute(DatabaseQuery.java:913)
	at org.eclipse.persistence.queries.ObjectLevelReadQuery.execute(ObjectLevelReadQuery.java:1195)
	at org.eclipse.persistence.queries.ReadAllQuery.execute(ReadAllQuery.java:485)
	at org.eclipse.persistence.queries.ObjectLevelReadQuery.executeInUnitOfWork(ObjectLevelReadQuery.java:1283)
	at org.eclipse.persistence.internal.sessions.UnitOfWorkImpl.internalExecuteQuery(UnitOfWorkImpl.java:3031)
	at org.eclipse.persistence.internal.sessions.AbstractSession.executeQuery(AbstractSession.java:1841)
	at org.eclipse.persistence.internal.sessions.AbstractSession.retryQuery(AbstractSession.java:1912)
	at org.eclipse.persistence.sessions.server.ClientSession.retryQuery(ClientSession.java:704)
	at org.eclipse.persistence.internal.sessions.UnitOfWorkImpl.retryQuery(UnitOfWorkImpl.java:5836)
	at org.eclipse.persistence.internal.sessions.AbstractSession.executeQuery(AbstractSession.java:1878)
	at org.eclipse.persistence.internal.sessions.AbstractSession.executeQuery(AbstractSession.java:1823)
	at org.eclipse.persistence.internal.sessions.AbstractSession.executeQuery(AbstractSession.java:1788)
	at org.eclipse.persistence.internal.jpa.QueryImpl.executeReadQuery(QueryImpl.java:263)
	at org.eclipse.persistence.internal.jpa.QueryImpl.getSingleResult(QueryImpl.java:531)
	at org.eclipse.persistence.internal.jpa.EJBQueryImpl.getSingleResult(EJBQueryImpl.java:400)
	at uep.game.lottery.draw.business.DrawManager.getNussPayinDrawByGameIdAndDrawNo(DrawManager.java:144)
	at uep.game.lottery.draw.business.DrawManager.getNussPayinDrawByGameIdAndDrawNoAsyncForCache(DrawManager.java:106)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at org.glassfish.exousia.AuthorizationService.lambda$invokeBeanMethod$1(AuthorizationService.java:506)
	at org.glassfish.exousia.AuthorizationService.runInScope(AuthorizationService.java:586)
	at org.glassfish.exousia.AuthorizationService.invokeBeanMethod(AuthorizationService.java:506)
	at org.glassfish.ejb.security.application.EJBSecurityManager.invoke(EJBSecurityManager.java:297)
	at com.sun.ejb.containers.BaseContainer.invokeBeanMethod(BaseContainer.java:4193)
	at com.sun.ejb.EjbInvocation.invokeBeanMethod(EjbInvocation.java:604)
	at com.sun.ejb.containers.interceptors.AroundInvokeChainImpl.invokeNext(InterceptorManager.java:706)
	at com.sun.ejb.EjbInvocation.proceed(EjbInvocation.java:553)
	at org.jboss.weld.module.ejb.AbstractEJBRequestScopeActivationInterceptor.aroundInvoke(AbstractEJBRequestScopeActivationInterceptor.java:87)
	at org.jboss.weld.module.ejb.SessionBeanInterceptor.aroundInvoke(SessionBeanInterceptor.java:52)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at com.sun.ejb.containers.interceptors.AroundInvokeInterceptor.intercept(InterceptorManager.java:766)
	at com.sun.ejb.containers.interceptors.AroundInvokeChainImpl.invokeNext(InterceptorManager.java:706)
	at com.sun.ejb.EjbInvocation.proceed(EjbInvocation.java:553)
	at com.sun.ejb.containers.interceptors.SystemInterceptorProxy.doCall(SystemInterceptorProxy.java:122)
	at com.sun.ejb.containers.interceptors.SystemInterceptorProxy.aroundInvoke(SystemInterceptorProxy.java:99)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at com.sun.ejb.containers.interceptors.AroundInvokeInterceptor.intercept(InterceptorManager.java:766)
	at com.sun.ejb.containers.interceptors.AroundInvokeChainImpl.invokeNext(InterceptorManager.java:706)
	at com.sun.ejb.containers.interceptors.InterceptorManager.intercept(InterceptorManager.java:307)
	at com.sun.ejb.containers.BaseContainer.__intercept(BaseContainer.java:4168)
	at com.sun.ejb.containers.BaseContainer.intercept(BaseContainer.java:4158)
	at com.sun.ejb.containers.EjbAsyncTask.call(EjbAsyncTask.java:75)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:317)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
	at java.base/java.lang.Thread.run(Thread.java:1583)
Caused by: oracle.net.ns.NetException: ORA-03113: database connection closed by peer (connection_type=client, source=/192.168.56.208:45924, target=rwdbuep2tfm1.lottery.co.at:1522, protocol=TCP, service=UEP2TFM1, detecting_end=client, server_type=unknown, client_program=GlassFishMain, server_process=29284, session=790.28240, user=UEP2_FK7, last_rpc=Execute query, connection_id=qjR4TE3tStqL3IXLUK3icA==, nne_encryption=null, nne_checksumming=null, authentication=O5LOGON)
https://docs.oracle.com/error-help/db/ora-03113/
	at oracle.jdbc.driver.T4CMAREngineNIO.prepareForUnmarshall(T4CMAREngineNIO.java:1004)
	at oracle.jdbc.driver.T4CMAREngineNIO.unmarshalUB1(T4CMAREngineNIO.java:466)
	at oracle.jdbc.driver.T4CTTIfun.receive(T4CTTIfun.java:814)
	at oracle.jdbc.driver.T4CTTIfun.doRPC(T4CTTIfun.java:237)
	at oracle.jdbc.driver.T4C8Oall.doOALL(T4C8Oall.java:516)
	at oracle.jdbc.driver.T4CPreparedStatement.doOall8(T4CPreparedStatement.java:275)
	at oracle.jdbc.driver.T4CPreparedStatement.executeForDescribe(T4CPreparedStatement.java:1232)
	... 134 more
Caused by: oracle.net.ns.NetException: ORA-17800: Got minus one from a read call. (CONNECTION_ID=qjR4TE3tStqL3IXLUK3icA==)
https://docs.oracle.com/error-help/db/ora-17800/
	at oracle.net.ns.NSProtocolNIO.doSocketRead(NSProtocolNIO.java:1227)
	at oracle.net.ns.NIOPacket.readHeader(NIOPacket.java:273)
	at oracle.net.ns.NIOPacket.readPacketFromSocketChannel(NIOPacket.java:206)
	at oracle.net.ns.NIOPacket.readFromSocketChannel(NIOPacket.java:149)
	at oracle.net.ns.NIOPacket.readFromSocketChannel(NIOPacket.java:122)
	at oracle.net.ns.NIONSDataChannel.readDataFromSocketChannel(NIONSDataChannel.java:112)
	at oracle.jdbc.driver.T4CMAREngineNIO.prepareForUnmarshall(T4CMAREngineNIO.java:932)
	... 140 more
]]

[2025-04-03T05:20:06.420000+02:00] [GF 7.0.22] [INFO] [] [oracle.jdbc] [tid: _ThreadID=224 _ThreadName=__ejb-thread-pool1] [levelValue: 800] [[
  properties={LOCALE=en_US, DriverVersion=23.6.0.24.10, java.library.path: =/server/glassfish7/glassfish/lib:/freigabe/bin/sqlplus_12c_64bit:/usr/java/packages/lib:/usr/lib64:/lib64:/lib:/usr/lib, java.class.path: =/server/glassfish7/glassfish/modules/glassfish.jar, DatabaseProductVersion=19000, java.version: =21.0.6, URL=jdbc:oracle:thin:@rwdbuep2tfm1.lottery.co.at:1522/UEP2TFM1}. ]]

[2025-04-03T05:20:04.964769+02:00] [GF 7.0.22] [INFO] [jdbc.exc_table_validation] [jakarta.enterprise.resource.resourceadapter.com.sun.gjc.spi] [tid: _ThreadID=224 _ThreadName=__ejb-thread-pool1] [levelValue: 800] [[
  RAR5074 : Table based validation detected invalid connection. Querying the table dual failed. Set resource-adapter log-level to FINE for exception stack trace.]]

[2025-04-03T05:10:05.234000+02:00] [GF 7.0.22] [INFO] [] [oracle.jdbc] [tid: _ThreadID=230 _ThreadName=__ejb-thread-pool7] [levelValue: 800] [[
  entering args (oracle.jdbc.datasource.impl.OracleDataSource$1@6f5248c1)]]

[2025-04-03T05:10:05.235000+02:00] [GF 7.0.22] [INFO] [] [oracle.jdbc] [tid: _ThreadID=230 _ThreadName=__ejb-thread-pool7] [levelValue: 800] [[
  traceId=1E5763B7. ]]

[2025-04-03T05:10:05.236000+02:00] [GF 7.0.22] [INFO] [] [oracle.jdbc] [tid: _ThreadID=230 _ThreadName=__ejb-thread-pool7] [levelValue: 800] [[
  Session Attributes: 
sdu=8192, tdu=2097152
nt: host=rwdbuep2tfm1.lottery.co.at, port=1522, socketOptions={0=YES, 1=NO, 2=20000, 38=TLS, 40=false, 8=/server/glassfish7/glassfish/domains/uepdomain/config/keystore.jks, 10=oracle.jdbc.internal.OpaqueString@5560f4db, 11=/server/glassfish7/glassfish/domains/uepdomain/config/cacerts.jks, 13=oracle.jdbc.internal.OpaqueString@5560f4db, 17=0, 18=false, 20=true, 23=40, 24=50, 25=0}
    socket=null
client profile={oracle.net.encryption_types_client=(), oracle.net.crypto_seed=, oracle.net.authentication_services=(), oracle.net.setFIPSMode=false, oracle.net.kerberos5_mutual_authentication=false, oracle.net.encryption_client=ACCEPTED, oracle.net.crypto_checksum_client=ACCEPTED, oracle.net.crypto_checksum_types_client=()}
connection options=[host=rwdbuep2tfm1.lottery.co.at port=1522 protocol=TCP service_name=UEP2TFM1 addr=(ADDRESS=(PROTOCOL=TCP)(HOST=rwdbuep2tfm1.lottery.co.at)(PORT=1522)) conn_data=(DESCRIPTION=(ADDRESS=(PROTOCOL=TCP)(PORT=1522)(HOST=192.168.53.68)(HOSTNAME=rwdbuep2tfm1.lottery.co.at))(CONNECT_DATA=(CID=(PROGRAM=GlassFishMain)(HOST=rwueptst7.lottery.co.at)(USER=bea))(SERVICE_NAME=UEP2TFM1))) done=true]
onBreakReset=false, dataEOF=false, negotiatedOptions=0xc41, connected=false
TTIINIT enabled=false, TTC cookie enabled=false
]]

[2025-04-03T05:10:05.237000+02:00] [GF 7.0.22] [INFO] [] [oracle.jdbc] [tid: _ThreadID=230 _ThreadName=__ejb-thread-pool7] [levelValue: 800] [[
  traceId=1E5763B7, anoEnabled=true. ]]

[2025-04-03T05:10:05.240000+02:00] [GF 7.0.22] [INFO] [] [oracle.jdbc] [tid: _ThreadID=230 _ThreadName=__ejb-thread-pool7] [levelValue: 800] [[
  Got Resend, SessionTraceId = 1E5763B7]]

[2025-04-03T05:10:05.243000+02:00] [GF 7.0.22] [INFO] [] [oracle.jdbc] [tid: _ThreadID=230 _ThreadName=__ejb-thread-pool7] [levelValue: 800] [[
  Connection established. Cleared conn strategy stack]]

[2025-04-03T05:10:05.245000+02:00] [GF 7.0.22] [INFO] [] [oracle.jdbc] [tid: _ThreadID=230 _ThreadName=__ejb-thread-pool7] [levelValue: 800] [[
  Sending break marker, SessionTraceId = 1E5763B7]]

[2025-04-03T05:10:05.256000+02:00] [GF 7.0.22] [INFO] [] [oracle.jdbc] [tid: _ThreadID=230 _ThreadName=__ejb-thread-pool7] [levelValue: 800] [[
  returning void]]

[2025-04-03T05:15:06.927000+02:00] [GF 7.0.22] [INFO] [] [oracle.jdbc] [tid: _ThreadID=215 _ThreadName=__ejb-thread-pool7] [levelValue: 800] [[
  SO_TIMEOUT interrupt timer cancelled null]]

[2025-04-03T05:15:06.928000+02:00] [GF 7.0.22] [INFO] [] [oracle.jdbc] [tid: _ThreadID=215 _ThreadName=__ejb-thread-pool7] [levelValue: 800] [[
  Sending urgent marker.  SessionTraceId=1E5763B7. ]]

[2025-04-03T05:20:06.427000+02:00] [GF 7.0.22] [INFO] [] [oracle.jdbc] [tid: _ThreadID=230 _ThreadName=__ejb-thread-pool7] [levelValue: 800] [[
  java.sql.SQLRecoverableException: ORA-03113: database connection closed by peer (connection_type=client, source=/192.168.56.208:45962, target=rwdbuep2tfm1.lottery.co.at:1522, protocol=TCP, service=UEP2TFM1, detecting_end=client, server_type=unknown, client_program=GlassFishMain, server_process=29303, session=862.10173, user=UEP2_FK7, last_rpc=Execute query, connection_id=6afFHeU/ScSHaJ373hdAqA==, nne_encryption=null, nne_checksumming=null, authentication=O5LOGON)
https://docs.oracle.com/error-help/db/ora-03113/
	at oracle.jdbc.driver.T4CPreparedStatement.executeForDescribe(T4CPreparedStatement.java:1246)
	at oracle.jdbc.driver.OracleStatement.prepareDefineBufferAndExecute(OracleStatement.java:1413)
	at oracle.jdbc.driver.OracleStatement.executeMaybeDescribe(OracleStatement.java:1287)
	at oracle.jdbc.driver.OracleStatement.executeSQLSelect(OracleStatement.java:1844)
	at oracle.jdbc.driver.OracleStatement.doExecuteWithTimeout(OracleStatement.java:1620)
	at oracle.jdbc.driver.OraclePreparedStatement.executeInternal(OraclePreparedStatement.java:3961)
	at oracle.jdbc.driver.OraclePreparedStatement.executeQuery(OraclePreparedStatement.java:4148)
	at oracle.jdbc.driver.OraclePreparedStatementWrapper.executeQuery(OraclePreparedStatementWrapper.java:1117)
	at com.sun.gjc.spi.jdbc40.PreparedStatementWrapper40.executeQuery(PreparedStatementWrapper40.java:656)
	at org.eclipse.persistence.internal.databaseaccess.DatabaseAccessor.executeSelect(DatabaseAccessor.java:1026)
	at org.eclipse.persistence.internal.databaseaccess.DatabaseAccessor.basicExecuteCall(DatabaseAccessor.java:662)
	at org.eclipse.persistence.internal.databaseaccess.DatabaseAccessor.executeCall(DatabaseAccessor.java:569)
	at org.eclipse.persistence.internal.sessions.AbstractSession.basicExecuteCall(AbstractSession.java:2048)
	at org.eclipse.persistence.sessions.server.ServerSession.executeCall(ServerSession.java:611)
	at org.eclipse.persistence.sessions.server.ClientSession.executeCall(ClientSession.java:263)
	at org.eclipse.persistence.internal.queries.DatasourceCallQueryMechanism.executeCall(DatasourceCallQueryMechanism.java:280)
	at org.eclipse.persistence.internal.queries.DatasourceCallQueryMechanism.executeCall(DatasourceCallQueryMechanism.java:266)
	at org.eclipse.persistence.internal.queries.DatasourceCallQueryMechanism.executeSelectCall(DatasourceCallQueryMechanism.java:352)
	at org.eclipse.persistence.internal.queries.DatasourceCallQueryMechanism.selectAllRows(DatasourceCallQueryMechanism.java:792)
	at org.eclipse.persistence.internal.queries.ExpressionQueryMechanism.selectAllRowsFromTable(ExpressionQueryMechanism.java:2848)
	at org.eclipse.persistence.internal.queries.ExpressionQueryMechanism.selectAllRows(ExpressionQueryMechanism.java:2801)
	at org.eclipse.persistence.queries.ReadAllQuery.executeObjectLevelReadQuery(ReadAllQuery.java:584)
	at org.eclipse.persistence.queries.ObjectLevelReadQuery.executeDatabaseQuery(ObjectLevelReadQuery.java:1236)
	at org.eclipse.persistence.queries.DatabaseQuery.execute(DatabaseQuery.java:913)
	at org.eclipse.persistence.queries.ObjectLevelReadQuery.execute(ObjectLevelReadQuery.java:1195)
	at org.eclipse.persistence.queries.ReadAllQuery.execute(ReadAllQuery.java:485)
	at org.eclipse.persistence.queries.ObjectLevelReadQuery.executeInUnitOfWork(ObjectLevelReadQuery.java:1283)
	at org.eclipse.persistence.internal.sessions.UnitOfWorkImpl.internalExecuteQuery(UnitOfWorkImpl.java:3031)
	at org.eclipse.persistence.internal.sessions.AbstractSession.executeQuery(AbstractSession.java:1841)
	at org.eclipse.persistence.internal.sessions.AbstractSession.retryQuery(AbstractSession.java:1912)
	at org.eclipse.persistence.sessions.server.ClientSession.retryQuery(ClientSession.java:704)
	at org.eclipse.persistence.internal.sessions.UnitOfWorkImpl.retryQuery(UnitOfWorkImpl.java:5836)
	at org.eclipse.persistence.internal.sessions.AbstractSession.executeQuery(AbstractSession.java:1878)
	at org.eclipse.persistence.internal.sessions.AbstractSession.executeQuery(AbstractSession.java:1823)
	at org.eclipse.persistence.internal.indirection.NoIndirectionPolicy.valueFromQuery(NoIndirectionPolicy.java:352)
	at org.eclipse.persistence.mappings.ForeignReferenceMapping.valueFromRowInternal(ForeignReferenceMapping.java:2343)
	at org.eclipse.persistence.mappings.ForeignReferenceMapping.valueFromRow(ForeignReferenceMapping.java:2184)
	at org.eclipse.persistence.mappings.ForeignReferenceMapping.buildCloneFromRow(ForeignReferenceMapping.java:341)
	at org.eclipse.persistence.internal.descriptors.ObjectBuilder.buildAttributesIntoWorkingCopyClone(ObjectBuilder.java:2111)
	at org.eclipse.persistence.internal.descriptors.ObjectBuilder.buildWorkingCopyCloneFromRow(ObjectBuilder.java:2364)
	at org.eclipse.persistence.internal.descriptors.ObjectBuilder.buildObjectInUnitOfWork(ObjectBuilder.java:958)
	at org.eclipse.persistence.internal.descriptors.ObjectBuilder.buildObjectInternal(ObjectBuilder.java:844)
	at org.eclipse.persistence.internal.descriptors.ObjectBuilder.buildObject(ObjectBuilder.java:799)
	at org.eclipse.persistence.internal.descriptors.ObjectBuilder.buildObject(ObjectBuilder.java:777)
	at org.eclipse.persistence.queries.ObjectLevelReadQuery.buildObject(ObjectLevelReadQuery.java:863)
	at org.eclipse.persistence.queries.ReadAllQuery.registerResultInUnitOfWork(ReadAllQuery.java:987)
	at org.eclipse.persistence.queries.ReadAllQuery.executeObjectLevelReadQuery(ReadAllQuery.java:598)
	at org.eclipse.persistence.queries.ObjectLevelReadQuery.executeDatabaseQuery(ObjectLevelReadQuery.java:1236)
	at org.eclipse.persistence.queries.DatabaseQuery.execute(DatabaseQuery.java:913)
	at org.eclipse.persistence.queries.ObjectLevelReadQuery.execute(ObjectLevelReadQuery.java:1195)
	at org.eclipse.persistence.queries.ReadAllQuery.execute(ReadAllQuery.java:485)
	at org.eclipse.persistence.queries.ObjectLevelReadQuery.executeInUnitOfWork(ObjectLevelReadQuery.java:1283)
	at org.eclipse.persistence.internal.sessions.UnitOfWorkImpl.internalExecuteQuery(UnitOfWorkImpl.java:3031)
	at org.eclipse.persistence.internal.sessions.AbstractSession.executeQuery(AbstractSession.java:1841)
	at org.eclipse.persistence.internal.sessions.AbstractSession.executeQuery(AbstractSession.java:1823)
	at org.eclipse.persistence.internal.sessions.AbstractSession.executeQuery(AbstractSession.java:1788)
	at org.eclipse.persistence.internal.jpa.QueryImpl.executeReadQuery(QueryImpl.java:263)
	at org.eclipse.persistence.internal.jpa.QueryImpl.getSingleResult(QueryImpl.java:531)
	at org.eclipse.persistence.internal.jpa.EJBQueryImpl.getSingleResult(EJBQueryImpl.java:400)
	at uep.game.lottery.draw.business.DrawManager.getNussPayinDrawByGameIdAndDrawNo(DrawManager.java:144)
	at uep.game.lottery.draw.business.DrawManager.getNussPayinDrawByGameIdAndDrawNoAsyncForCache(DrawManager.java:106)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at org.glassfish.exousia.AuthorizationService.lambda$invokeBeanMethod$1(AuthorizationService.java:506)
	at org.glassfish.exousia.AuthorizationService.runInScope(AuthorizationService.java:586)
	at org.glassfish.exousia.AuthorizationService.invokeBeanMethod(AuthorizationService.java:506)
	at org.glassfish.ejb.security.application.EJBSecurityManager.invoke(EJBSecurityManager.java:297)
	at com.sun.ejb.containers.BaseContainer.invokeBeanMethod(BaseContainer.java:4193)
	at com.sun.ejb.EjbInvocation.invokeBeanMethod(EjbInvocation.java:604)
	at com.sun.ejb.containers.interceptors.AroundInvokeChainImpl.invokeNext(InterceptorManager.java:706)
	at com.sun.ejb.EjbInvocation.proceed(EjbInvocation.java:553)
	at org.jboss.weld.module.ejb.AbstractEJBRequestScopeActivationInterceptor.aroundInvoke(AbstractEJBRequestScopeActivationInterceptor.java:87)
	at org.jboss.weld.module.ejb.SessionBeanInterceptor.aroundInvoke(SessionBeanInterceptor.java:52)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at com.sun.ejb.containers.interceptors.AroundInvokeInterceptor.intercept(InterceptorManager.java:766)
	at com.sun.ejb.containers.interceptors.AroundInvokeChainImpl.invokeNext(InterceptorManager.java:706)
	at com.sun.ejb.EjbInvocation.proceed(EjbInvocation.java:553)
	at com.sun.ejb.containers.interceptors.SystemInterceptorProxy.doCall(SystemInterceptorProxy.java:122)
	at com.sun.ejb.containers.interceptors.SystemInterceptorProxy.aroundInvoke(SystemInterceptorProxy.java:99)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at com.sun.ejb.containers.interceptors.AroundInvokeInterceptor.intercept(InterceptorManager.java:766)
	at com.sun.ejb.containers.interceptors.AroundInvokeChainImpl.invokeNext(InterceptorManager.java:706)
	at com.sun.ejb.containers.interceptors.InterceptorManager.intercept(InterceptorManager.java:307)
	at com.sun.ejb.containers.BaseContainer.__intercept(BaseContainer.java:4168)
	at com.sun.ejb.containers.BaseContainer.intercept(BaseContainer.java:4158)
	at com.sun.ejb.containers.EjbAsyncTask.call(EjbAsyncTask.java:75)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:317)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
	at java.base/java.lang.Thread.run(Thread.java:1583)
Caused by: oracle.net.ns.NetException: ORA-03113: database connection closed by peer (connection_type=client, source=/192.168.56.208:45962, target=rwdbuep2tfm1.lottery.co.at:1522, protocol=TCP, service=UEP2TFM1, detecting_end=client, server_type=unknown, client_program=GlassFishMain, server_process=29303, session=862.10173, user=UEP2_FK7, last_rpc=Execute query, connection_id=6afFHeU/ScSHaJ373hdAqA==, nne_encryption=null, nne_checksumming=null, authentication=O5LOGON)
https://docs.oracle.com/error-help/db/ora-03113/
	at oracle.jdbc.driver.T4CMAREngineNIO.prepareForUnmarshall(T4CMAREngineNIO.java:1004)
	at oracle.jdbc.driver.T4CMAREngineNIO.unmarshalUB1(T4CMAREngineNIO.java:466)
	at oracle.jdbc.driver.T4CTTIfun.receive(T4CTTIfun.java:814)
	at oracle.jdbc.driver.T4CTTIfun.doRPC(T4CTTIfun.java:237)
	at oracle.jdbc.driver.T4C8Oall.doOALL(T4C8Oall.java:516)
	at oracle.jdbc.driver.T4CPreparedStatement.doOall8(T4CPreparedStatement.java:275)
	at oracle.jdbc.driver.T4CPreparedStatement.executeForDescribe(T4CPreparedStatement.java:1232)
	... 91 more
Caused by: oracle.net.ns.NetException: ORA-17800: Got minus one from a read call. (CONNECTION_ID=6afFHeU/ScSHaJ373hdAqA==)
https://docs.oracle.com/error-help/db/ora-17800/
	at oracle.net.ns.NSProtocolNIO.doSocketRead(NSProtocolNIO.java:1227)
	at oracle.net.ns.NIOPacket.readHeader(NIOPacket.java:273)
	at oracle.net.ns.NIOPacket.readPacketFromSocketChannel(NIOPacket.java:206)
	at oracle.net.ns.NIOPacket.readFromSocketChannel(NIOPacket.java:149)
	at oracle.net.ns.NIOPacket.readFromSocketChannel(NIOPacket.java:122)
	at oracle.net.ns.NIONSDataChannel.readDataFromSocketChannel(NIONSDataChannel.java:112)
	at oracle.jdbc.driver.T4CMAREngineNIO.prepareForUnmarshall(T4CMAREngineNIO.java:932)
	... 97 more
]]

[2025-04-03T05:20:06.428000+02:00] [GF 7.0.22] [INFO] [] [oracle.jdbc] [tid: _ThreadID=230 _ThreadName=__ejb-thread-pool7] [levelValue: 800] [[
  properties={LOCALE=en_US, DriverVersion=23.6.0.24.10, java.library.path: =/server/glassfish7/glassfish/lib:/freigabe/bin/sqlplus_12c_64bit:/usr/java/packages/lib:/usr/lib64:/lib64:/lib:/usr/lib, java.class.path: =/server/glassfish7/glassfish/modules/glassfish.jar, DatabaseProductVersion=19000, java.version: =21.0.6, URL=jdbc:oracle:thin:@rwdbuep2tfm1.lottery.co.at:1522/UEP2TFM1}. ]]

[2025-04-03T05:10:05.194000+02:00] [GF 7.0.22] [INFO] [] [oracle.jdbc] [tid: _ThreadID=226 _ThreadName=__ejb-thread-pool3] [levelValue: 800] [[
  entering args (oracle.jdbc.datasource.impl.OracleDataSource$1@27e9db02)]]

[2025-04-03T05:10:05.195000+02:00] [GF 7.0.22] [INFO] [] [oracle.jdbc] [tid: _ThreadID=226 _ThreadName=__ejb-thread-pool3] [levelValue: 800] [[
  traceId=1CDDF9B2. ]]

[2025-04-03T05:10:05.196000+02:00] [GF 7.0.22] [INFO] [] [oracle.jdbc] [tid: _ThreadID=226 _ThreadName=__ejb-thread-pool3] [levelValue: 800] [[
  Session Attributes: 
sdu=8192, tdu=2097152
nt: host=rwdbuep2tfm1.lottery.co.at, port=1522, socketOptions={0=YES, 1=NO, 2=20000, 38=TLS, 40=false, 8=/server/glassfish7/glassfish/domains/uepdomain/config/keystore.jks, 10=oracle.jdbc.internal.OpaqueString@5560f4db, 11=/server/glassfish7/glassfish/domains/uepdomain/config/cacerts.jks, 13=oracle.jdbc.internal.OpaqueString@5560f4db, 17=0, 18=false, 20=true, 23=40, 24=50, 25=0}
    socket=null
client profile={oracle.net.encryption_types_client=(), oracle.net.crypto_seed=, oracle.net.authentication_services=(), oracle.net.setFIPSMode=false, oracle.net.kerberos5_mutual_authentication=false, oracle.net.encryption_client=ACCEPTED, oracle.net.crypto_checksum_client=ACCEPTED, oracle.net.crypto_checksum_types_client=()}
connection options=[host=rwdbuep2tfm1.lottery.co.at port=1522 protocol=TCP service_name=UEP2TFM1 addr=(ADDRESS=(PROTOCOL=TCP)(HOST=rwdbuep2tfm1.lottery.co.at)(PORT=1522)) conn_data=(DESCRIPTION=(ADDRESS=(PROTOCOL=TCP)(PORT=1522)(HOST=192.168.53.68)(HOSTNAME=rwdbuep2tfm1.lottery.co.at))(CONNECT_DATA=(CID=(PROGRAM=GlassFishMain)(HOST=rwueptst7.lottery.co.at)(USER=bea))(SERVICE_NAME=UEP2TFM1))) done=true]
onBreakReset=false, dataEOF=false, negotiatedOptions=0xc41, connected=false
TTIINIT enabled=false, TTC cookie enabled=false
]]

[2025-04-03T05:10:05.197000+02:00] [GF 7.0.22] [INFO] [] [oracle.jdbc] [tid: _ThreadID=226 _ThreadName=__ejb-thread-pool3] [levelValue: 800] [[
  traceId=1CDDF9B2, anoEnabled=true. ]]

[2025-04-03T05:10:05.200000+02:00] [GF 7.0.22] [INFO] [] [oracle.jdbc] [tid: _ThreadID=226 _ThreadName=__ejb-thread-pool3] [levelValue: 800] [[
  Got Resend, SessionTraceId = 1CDDF9B2]]

[2025-04-03T05:10:05.203000+02:00] [GF 7.0.22] [INFO] [] [oracle.jdbc] [tid: _ThreadID=226 _ThreadName=__ejb-thread-pool3] [levelValue: 800] [[
  Connection established. Cleared conn strategy stack]]

[2025-04-03T05:10:05.205000+02:00] [GF 7.0.22] [INFO] [] [oracle.jdbc] [tid: _ThreadID=226 _ThreadName=__ejb-thread-pool3] [levelValue: 800] [[
  Sending break marker, SessionTraceId = 1CDDF9B2]]

[2025-04-03T05:10:05.217000+02:00] [GF 7.0.22] [INFO] [] [oracle.jdbc] [tid: _ThreadID=226 _ThreadName=__ejb-thread-pool3] [levelValue: 800] [[
  returning void]]

[2025-04-03T05:15:06.929000+02:00] [GF 7.0.22] [INFO] [] [oracle.jdbc] [tid: _ThreadID=215 _ThreadName=__ejb-thread-pool3] [levelValue: 800] [[
  SO_TIMEOUT interrupt timer cancelled null]]

[2025-04-03T05:15:06.930000+02:00] [GF 7.0.22] [INFO] [] [oracle.jdbc] [tid: _ThreadID=215 _ThreadName=__ejb-thread-pool3] [levelValue: 800] [[
  Sending urgent marker.  SessionTraceId=1CDDF9B2. ]]

[2025-04-03T05:20:06.430000+02:00] [GF 7.0.22] [INFO] [] [oracle.jdbc] [tid: _ThreadID=226 _ThreadName=__ejb-thread-pool3] [levelValue: 800] [[

--- a/appserver/connectors/connectors-runtime/src/main/java/com/sun/enterprise/resource/ConnectorXAResource.java
+++ b/appserver/connectors/connectors-runtime/src/main/java/com/sun/enterprise/resource/ConnectorXAResource.java
@@ -192,11 +192,16 @@ public class ConnectorXAResource implements XAResource {
         return false;
     }
 
-    private ResourceHandle getResourceHandle() throws PoolingException {
+    // package default for unit test
+    ResourceHandle getResourceHandle() throws PoolingException {
+        return getResourceHandle(true);
+    }
+
+    private ResourceHandle getResourceHandle(boolean beginLocalTransactionIfStateIsUnenlisted) throws PoolingException {
         try {
             ResourceHandle resourceHandle = null;
             JavaEETransaction javaEETransaction = getCurrentTransaction();
-            if (javaEETransaction == null) { // Only if some thing is wrong with tx manager.
+            if (javaEETransaction == null || javaEETransaction.getNonXAResource() == null) { // Only if some thing is wrong with tx manager.
                 resourceHandle = localHandle_; // Just return the local handle.
             } else {
                 resourceHandle = (ResourceHandle) javaEETransaction.getNonXAResource();
@@ -210,7 +215,7 @@ public class ConnectorXAResource implements XAResource {
                 }
             }
 
-            if (resourceHandle.getResourceState().isUnenlisted()) {
+            if (resourceHandle.getResourceState().isUnenlisted() && beginLocalTransactionIfStateIsUnenlisted) {
                 ManagedConnection managedConnection = resourceHandle.getResource();
 
                 // Begin the local transaction if first time
@@ -224,13 +229,16 @@ public class ConnectorXAResource implements XAResource {
         }
     }
 
-    private JavaEETransaction getCurrentTransaction() throws SystemException {
+    // package default for unit test
+    JavaEETransaction getCurrentTransaction() throws SystemException {
         return (JavaEETransaction) ConnectorRuntime.getRuntime().getTransactionManager().getTransaction();
     }
 
     private void resetAssociation() throws XAException {
         try {
-            ResourceHandle handle = getResourceHandle();
+            // Do not begin a local managed connection in case of reset association
+            ResourceHandle handle = getResourceHandle(false);
+
             LocalTxConnectionEventListener listener = (LocalTxConnectionEventListener) handle.getListener();
 
             // Clear the associations and Map all associated handles back to their actual Managed Connection.

--- a/appserver/connectors/connectors-runtime/src/test/java/com/sun/enterprise/resource/ConnectorXAResourceTest.java
+++ b/appserver/connectors/connectors-runtime/src/test/java/com/sun/enterprise/resource/ConnectorXAResourceTest.java
@@ -1,0 +1,432 @@
+/*
+ * Copyright (c) 2024 Eclipse Foundation and/or its affiliates. All rights reserved.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License v. 2.0, which is available at
+ * http://www.eclipse.org/legal/epl-2.0.
+ *
+ * This Source Code may also be made available under the following Secondary
+ * Licenses when the conditions for such availability set forth in the
+ * Eclipse Public License v. 2.0 are satisfied: GNU General Public License,
+ * version 2 with the GNU Classpath Exception, which is available at
+ * https://www.gnu.org/software/classpath/license.html.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+ */
+
+package com.sun.enterprise.resource;
+
+import com.sun.appserv.connectors.internal.api.ConnectorRuntimeException;
+import com.sun.enterprise.connectors.ConnectorRuntime;
+import com.sun.enterprise.resource.allocator.LocalTxConnectorAllocator;
+import com.sun.enterprise.resource.allocator.ResourceAllocator;
+import com.sun.enterprise.resource.pool.InjectionUtil;
+import com.sun.enterprise.resource.pool.mock.JavaEETransactionManagerMock;
+import com.sun.enterprise.resource.pool.mock.JavaEETransactionMock;
+import com.sun.enterprise.resource.pool.mock.LocalTransactionMock;
+import com.sun.enterprise.resource.pool.mock.ManagedConnectionFactoryMock;
+import com.sun.enterprise.resource.pool.mock.ManagedConnectionMock;
+import com.sun.enterprise.transaction.api.JavaEETransactionManager;
+import com.sun.enterprise.transaction.spi.TransactionalResource;
+
+import jakarta.resource.ResourceException;
+import jakarta.resource.spi.ConnectionRequestInfo;
+import jakarta.resource.spi.LocalTransaction;
+import jakarta.resource.spi.ManagedConnection;
+import jakarta.resource.spi.ManagedConnectionFactory;
+import jakarta.transaction.SystemException;
+import jakarta.transaction.Transaction;
+
+import javax.security.auth.Subject;
+
+import org.glassfish.api.admin.ProcessEnvironment;
+import org.glassfish.api.naming.SimpleJndiName;
+import org.glassfish.resourcebase.resources.api.PoolInfo;
+import org.junit.jupiter.api.Test;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertFalse;
+import static org.junit.jupiter.api.Assertions.assertNotNull;
+import static org.junit.jupiter.api.Assertions.assertNull;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+public class ConnectorXAResourceTest {
+
+    private ResourceSpec resourceSpec = new ResourceSpec(new SimpleJndiName("myResourceSpec"), ResourceSpec.JNDI_NAME);
+
+    @Test
+    void testGetResourceHandle_CurrentTransaction_Null() throws Exception {
+        // Boilerplate
+        MyLocalTransactionMock localTransaction = new MyLocalTransactionMock();
+        ManagedConnection managedConnection = new MyManagedConnection(localTransaction);
+        ManagedConnectionFactory managedConnectionFactory = new MyManagedConnectionFactory(managedConnection);
+
+        // Essence of the test: no transaction
+        MyJavaEETransaction javaEETransaction = null;
+
+        // Make sure ConnectorRuntime singleton is initialized
+        MyConnectorRuntime connectorRuntime = new MyConnectorRuntime(new MyJavaEETransactionManager(javaEETransaction));
+        connectorRuntime.postConstruct();
+
+        // Create ResourceAllocator
+        ResourceAllocator allocator = new LocalTxConnectorAllocator(null, managedConnectionFactory, resourceSpec, null,
+                null, null, null, false);
+
+        // Create ResourceHandle
+        ResourceHandle handle = allocator.createResource();
+        ResourceSpec resourceSpec = handle.getResourceSpec();
+        ResourceAllocator resourceAllocator = handle.getResourceAllocator();
+        ClientSecurityInfo info = null;
+
+        // Create ConnectorXAResource for the handle
+        ConnectorXAResource connectorXAResource = new ConnectorXAResource(handle, resourceSpec, resourceAllocator, info);
+        // Purpose of this test variation: current transaction is null
+        assertNull(connectorXAResource.getCurrentTransaction());
+
+        // Test getResource while current transaction is null
+        ResourceHandle resourceHandle = connectorXAResource.getResourceHandle();
+        assertEquals(handle, resourceHandle);
+
+        // Make sure the resource is unenlisted
+        assertFalse(resourceHandle.isEnlisted());
+        assertTrue(resourceHandle.getResourceState().isUnenlisted());
+
+        // In case the resource was unenlisted the transaction should also have been started
+        LocalTransaction transaction = resourceHandle.getResource().getLocalTransaction();
+        assertEquals(localTransaction, transaction);
+        assertEquals(1, localTransaction.getBeginNrOfCalls());
+    }
+
+    @Test
+    void testGetResourceHandle_NonXaResource_Null() throws Exception {
+        // Boilerplate
+        MyLocalTransactionMock localTransaction = new MyLocalTransactionMock();
+        ManagedConnection managedConnection = new MyManagedConnection(localTransaction);
+        ManagedConnectionFactory managedConnectionFactory = new MyManagedConnectionFactory(managedConnection);
+
+        // Essence of the test MyJavaEETransaction returns null for getNonXaResource
+        MyJavaEETransaction javaEETransaction = new MyJavaEETransaction();
+        assertNull(javaEETransaction.getNonXAResource());
+
+        // Make sure ConnectorRuntime singleton is initialized
+        MyConnectorRuntime connectorRuntime = new MyConnectorRuntime(new MyJavaEETransactionManager(javaEETransaction));
+        connectorRuntime.postConstruct();
+
+        // Create ResourceAllocator
+        ResourceAllocator allocator = new LocalTxConnectorAllocator(null, managedConnectionFactory, resourceSpec, null,
+                null, null, null, false);
+
+        // Create ResourceHandle
+        ResourceHandle handle = allocator.createResource();
+        ResourceSpec resourceSpec = handle.getResourceSpec();
+        ResourceAllocator resourceAllocator = handle.getResourceAllocator();
+        ClientSecurityInfo info = null;
+
+        // Create ConnectorXAResource for the handle
+        ConnectorXAResource connectorXAResource = new ConnectorXAResource(handle, resourceSpec, resourceAllocator, info);
+        // Purpose of this test variation: current transaction is not null, but NonXaResource is null
+        assertNotNull(connectorXAResource.getCurrentTransaction());
+        assertNull(connectorXAResource.getCurrentTransaction().getNonXAResource());
+        assertEquals(2, javaEETransaction.getGetNonXAResourceNrOfCalls());
+
+        // Before the getResourceHandle, check the transaction is not yet started
+        assertEquals(0, localTransaction.getBeginNrOfCalls());
+
+        // Test getResource while NonXaResource is null
+        ResourceHandle resourceHandle = connectorXAResource.getResourceHandle();
+        assertEquals(handle, resourceHandle);
+        assertEquals(3, javaEETransaction.getGetNonXAResourceNrOfCalls());
+
+        // Make sure the resource is unenlisted
+        assertFalse(resourceHandle.isEnlisted());
+        assertTrue(resourceHandle.getResourceState().isUnenlisted());
+
+        // In case the resource was unenlisted the transaction should also have been started
+        LocalTransaction transaction = resourceHandle.getResource().getLocalTransaction();
+        assertEquals(localTransaction, transaction);
+        assertEquals(1, localTransaction.getBeginNrOfCalls());
+    }
+
+    @Test
+    void testRollback_CurrentTransaction_Null_Unenlisted() throws Exception {
+        // Boilerplate
+        MyLocalTransactionMock localTransaction = new MyLocalTransactionMock();
+        ManagedConnection managedConnection = new MyManagedConnection(localTransaction);
+        ManagedConnectionFactory managedConnectionFactory = new MyManagedConnectionFactory(managedConnection);
+
+        // Essence of the test: no transaction
+        MyJavaEETransaction javaEETransaction = null;
+
+        // Make sure ConnectorRuntime singleton is initialized
+        MyConnectorRuntime connectorRuntime = new MyConnectorRuntime(new MyJavaEETransactionManager(javaEETransaction));
+        connectorRuntime.postConstruct();
+
+        // Create ResourceAllocator
+        ResourceAllocator allocator = new LocalTxConnectorAllocator(null, managedConnectionFactory, resourceSpec, null,
+                null, null, null, false);
+
+        // Create ResourceHandle
+        ResourceHandle resourceHandle = allocator.createResource();
+        ResourceSpec resourceSpec = resourceHandle.getResourceSpec();
+        ResourceAllocator resourceAllocator = resourceHandle.getResourceAllocator();
+        ClientSecurityInfo info = null;
+
+        // Create ConnectorXAResource for the handle
+        ConnectorXAResource connectorXAResource = new ConnectorXAResource(resourceHandle, resourceSpec, resourceAllocator, info);
+        // Purpose of this test variation: current transaction is null
+        assertNull(connectorXAResource.getCurrentTransaction());
+
+        // Test the rollback
+        connectorXAResource.rollback(/* xidNotRequired */ null);
+
+        // Make sure the resource is unenlisted
+        assertFalse(resourceHandle.isEnlisted());
+        assertTrue(resourceHandle.getResourceState().isUnenlisted());
+
+        // In case the resource was unenlisted the transaction should also have been started
+        LocalTransaction transaction = resourceHandle.getResource().getLocalTransaction();
+        assertEquals(localTransaction, transaction);
+        // Value should be 1 due to getResourceHandle calling begin for unenlisted resource
+        // It should not be 2 due to resetAssociation calling begin again
+        assertEquals(1, localTransaction.getBeginNrOfCalls());
+
+        // Rollback should have erased the associated transaction
+        assertNull(connectorXAResource.getAssociatedTransaction());
+    }
+
+    @Test
+    void testRollback_CurrentTransaction_Null_Enlisted() throws Exception {
+        // Boilerplate
+        MyLocalTransactionMock localTransaction = new MyLocalTransactionMock();
+        ManagedConnection managedConnection = new MyManagedConnection(localTransaction);
+        ManagedConnectionFactory managedConnectionFactory = new MyManagedConnectionFactory(managedConnection);
+
+        // Essence of the test: no transaction
+        MyJavaEETransaction javaEETransaction = null;
+
+        // Make sure ConnectorRuntime singleton is initialized
+        MyConnectorRuntime connectorRuntime = new MyConnectorRuntime(new MyJavaEETransactionManager(javaEETransaction));
+        connectorRuntime.postConstruct();
+
+        // Create ResourceAllocator
+        ResourceAllocator allocator = new LocalTxConnectorAllocator(null, managedConnectionFactory, resourceSpec, null,
+                null, null, null, false);
+
+        // Create ResourceHandle
+        ResourceHandle resourceHandle = allocator.createResource();
+        ResourceSpec resourceSpec = resourceHandle.getResourceSpec();
+        ResourceAllocator resourceAllocator = resourceHandle.getResourceAllocator();
+        ClientSecurityInfo info = null;
+
+        // Create ConnectorXAResource for the handle
+        ConnectorXAResource connectorXAResource = new ConnectorXAResource(resourceHandle, resourceSpec, resourceAllocator, info);
+        // Purpose of this test variation: current transaction is null
+        assertNull(connectorXAResource.getCurrentTransaction());
+
+        // Make sure the resource is enlisted
+        resourceHandle.getResourceState().setEnlisted(true);
+
+        // Test the rollback
+        connectorXAResource.rollback(/* xidNotRequired */ null);
+
+        // Make sure the resource is still enlisted
+        assertTrue(resourceHandle.isEnlisted());
+        assertTrue(resourceHandle.getResourceState().isEnlisted());
+
+        // In case the resource was unenlisted the transaction should also have been started
+        LocalTransaction transaction = resourceHandle.getResource().getLocalTransaction();
+        assertEquals(localTransaction, transaction);
+        // Value should be 0, no begin should have been called
+        assertEquals(0, localTransaction.getBeginNrOfCalls());
+
+        // Rollback should have erased the associated transaction
+        assertNull(connectorXAResource.getAssociatedTransaction());
+    }
+
+    @Test
+    void testRollback_NonXaResource_Null_Unenlisted() throws Exception {
+        // Boilerplate
+        MyLocalTransactionMock localTransaction = new MyLocalTransactionMock();
+        ManagedConnection managedConnection = new MyManagedConnection(localTransaction);
+        ManagedConnectionFactory managedConnectionFactory = new MyManagedConnectionFactory(managedConnection);
+
+        // Essence of the test MyJavaEETransaction returns null for getNonXaResource
+        MyJavaEETransaction javaEETransaction = new MyJavaEETransaction();
+        assertNull(javaEETransaction.getNonXAResource());
+
+        // Make sure ConnectorRuntime singleton is initialized
+        MyConnectorRuntime connectorRuntime = new MyConnectorRuntime(new MyJavaEETransactionManager(javaEETransaction));
+        connectorRuntime.postConstruct();
+
+        // Create ResourceAllocator
+        ResourceAllocator allocator = new LocalTxConnectorAllocator(null, managedConnectionFactory, resourceSpec, null,
+                null, null, null, false);
+
+        // Create ResourceHandle
+        ResourceHandle resourceHandle = allocator.createResource();
+        ResourceSpec resourceSpec = resourceHandle.getResourceSpec();
+        ResourceAllocator resourceAllocator = resourceHandle.getResourceAllocator();
+        ClientSecurityInfo info = null;
+
+        // Create ConnectorXAResource for the handle
+        ConnectorXAResource connectorXAResource = new ConnectorXAResource(resourceHandle, resourceSpec, resourceAllocator, info);
+        // Purpose of this test variation: current transaction is not null, but NonXaResource is null
+        assertNotNull(connectorXAResource.getCurrentTransaction());
+        assertNull(connectorXAResource.getCurrentTransaction().getNonXAResource());
+        assertEquals(2, javaEETransaction.getGetNonXAResourceNrOfCalls());
+
+        // Test the rollback
+        connectorXAResource.rollback(/* xidNotRequired */ null);
+
+        // Make sure the resource is unenlisted
+        assertFalse(resourceHandle.isEnlisted());
+        assertTrue(resourceHandle.getResourceState().isUnenlisted());
+
+        // In case the resource was unenlisted the transaction should also have been started
+        LocalTransaction transaction = resourceHandle.getResource().getLocalTransaction();
+        assertEquals(localTransaction, transaction);
+        // Value should be 1 due to getResourceHandle calling begin for unenlisted resource
+        // It should not be 2 due to resetAssociation calling begin again
+        assertEquals(1, localTransaction.getBeginNrOfCalls());
+
+        // Rollback should have erased the associated transaction
+        assertNull(connectorXAResource.getAssociatedTransaction());
+    }
+
+    @Test
+    void testRollback_NonXaResource_Null_Enlisted() throws Exception {
+        // Boilerplate
+        MyLocalTransactionMock localTransaction = new MyLocalTransactionMock();
+        ManagedConnection managedConnection = new MyManagedConnection(localTransaction);
+        ManagedConnectionFactory managedConnectionFactory = new MyManagedConnectionFactory(managedConnection);
+
+        // Essence of the test MyJavaEETransaction returns null for getNonXaResource
+        MyJavaEETransaction javaEETransaction = new MyJavaEETransaction();
+        assertNull(javaEETransaction.getNonXAResource());
+
+        // Make sure ConnectorRuntime singleton is initialized
+        MyConnectorRuntime connectorRuntime = new MyConnectorRuntime(new MyJavaEETransactionManager(javaEETransaction));
+        connectorRuntime.postConstruct();
+
+        // Create ResourceAllocator
+        ResourceAllocator allocator = new LocalTxConnectorAllocator(null, managedConnectionFactory, resourceSpec, null,
+                null, null, null, false);
+
+        // Create ResourceHandle
+        ResourceHandle resourceHandle = allocator.createResource();
+        ResourceSpec resourceSpec = resourceHandle.getResourceSpec();
+        ResourceAllocator resourceAllocator = resourceHandle.getResourceAllocator();
+        ClientSecurityInfo info = null;
+
+        // Create ConnectorXAResource for the handle
+        ConnectorXAResource connectorXAResource = new ConnectorXAResource(resourceHandle, resourceSpec, resourceAllocator, info);
+        // Purpose of this test variation: current transaction is not null, but NonXaResource is null
+        assertNotNull(connectorXAResource.getCurrentTransaction());
+        assertNull(connectorXAResource.getCurrentTransaction().getNonXAResource());
+        assertEquals(2, javaEETransaction.getGetNonXAResourceNrOfCalls());
+
+        // Make sure the resource is enlisted
+        resourceHandle.getResourceState().setEnlisted(true);
+
+        // Test the rollback
+        connectorXAResource.rollback(/* xidNotRequired */ null);
+
+        // Make sure the resource is still enlisted
+        assertTrue(resourceHandle.isEnlisted());
+        assertTrue(resourceHandle.getResourceState().isEnlisted());
+
+        // In case the resource was unenlisted the transaction should also have been started
+        LocalTransaction transaction = resourceHandle.getResource().getLocalTransaction();
+        assertEquals(localTransaction, transaction);
+        // Value should be 0, no begin should have been called
+        assertEquals(0, localTransaction.getBeginNrOfCalls());
+
+        // Rollback should have erased the associated transaction
+        assertNull(connectorXAResource.getAssociatedTransaction());
+    }
+
+    public class MyManagedConnectionFactory extends ManagedConnectionFactoryMock {
+        private static final long serialVersionUID = 1L;
+        private ManagedConnection managedConnection;
+
+        public MyManagedConnectionFactory(ManagedConnection managedConnection) {
+            this.managedConnection = managedConnection;
+        }
+
+        @Override
+        public ManagedConnection createManagedConnection(Subject subject, ConnectionRequestInfo cxRequestInfo)
+                throws ResourceException {
+            return managedConnection;
+        }
+    }
+
+    public class MyManagedConnection extends ManagedConnectionMock {
+        private LocalTransaction localTransaction;
+
+        public MyManagedConnection(LocalTransaction localTransaction) {
+            this.localTransaction = localTransaction;
+        }
+
+        @Override
+        public LocalTransaction getLocalTransaction() throws ResourceException {
+            return localTransaction;
+        }
+    }
+
+    public class MyJavaEETransactionManager extends JavaEETransactionManagerMock {
+        private Transaction transaction;
+
+        public MyJavaEETransactionManager(Transaction transaction) {
+            this.transaction = transaction;
+        }
+
+        @Override
+        public Transaction getTransaction() throws SystemException {
+            return transaction;
+        }
+    }
+
+    private class MyJavaEETransaction extends JavaEETransactionMock {
+        public int getNonXAResourceNrOfCalls;
+
+        @Override
+        public TransactionalResource getNonXAResource() {
+            getNonXAResourceNrOfCalls = getGetNonXAResourceNrOfCalls() + 1;
+            // Make sure the return value is null
+            return null;
+        }
+
+        public int getGetNonXAResourceNrOfCalls() {
+            return getNonXAResourceNrOfCalls;
+        }
+    }
+
+    public class MyLocalTransactionMock extends LocalTransactionMock {
+        private int getBeginNrOfCalls;
+
+        @Override
+        public void begin() throws ResourceException {
+            getBeginNrOfCalls++;
+        }
+
+        public int getBeginNrOfCalls() {
+            return getBeginNrOfCalls;
+        }
+    }
+
+    public class MyConnectorRuntime extends ConnectorRuntime {
+        private ProcessEnvironment processEnvironment = new ProcessEnvironment();
+
+        public MyConnectorRuntime(JavaEETransactionManager transactionManager) throws Exception {
+            // Force 'injection' of private field processEnvironment
+            InjectionUtil.injectPrivateField(ConnectorRuntime.class, this, "processEnvironment", processEnvironment);
+            InjectionUtil.injectPrivateField(ConnectorRuntime.class, this, "transactionManager", transactionManager);
+        }
+
+        @Override
+        public PoolType getPoolType(PoolInfo poolInfo) throws ConnectorRuntimeException {
+            return PoolType.STANDARD_POOL;
+        }
+    }
+
+}

--- a/appserver/connectors/connectors-runtime/src/test/java/com/sun/enterprise/resource/pool/mock/LocalTransactionMock.java
+++ b/appserver/connectors/connectors-runtime/src/test/java/com/sun/enterprise/resource/pool/mock/LocalTransactionMock.java
@@ -1,0 +1,38 @@
+/*
+ * Copyright (c) 2025 Eclipse Foundation and/or its affiliates. All rights reserved.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License v. 2.0, which is available at
+ * http://www.eclipse.org/legal/epl-2.0.
+ *
+ * This Source Code may also be made available under the following Secondary
+ * Licenses when the conditions for such availability set forth in the
+ * Eclipse Public License v. 2.0 are satisfied: GNU General Public License,
+ * version 2 with the GNU Classpath Exception, which is available at
+ * https://www.gnu.org/software/classpath/license.html.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+ */
+
+package com.sun.enterprise.resource.pool.mock;
+
+import jakarta.resource.ResourceException;
+import jakarta.resource.spi.LocalTransaction;
+
+/**
+ * Mock class without any implementation
+ */
+public class LocalTransactionMock implements LocalTransaction {
+
+    @Override
+    public void begin() throws ResourceException {
+    }
+
+    @Override
+    public void commit() throws ResourceException {
+    }
+
+    @Override
+    public void rollback() throws ResourceException {
+    }
+}

--- a/appserver/connectors/connectors-runtime/src/test/java/com/sun/enterprise/resource/pool/mock/ManagedConnectionFactoryMock.java
+++ b/appserver/connectors/connectors-runtime/src/test/java/com/sun/enterprise/resource/pool/mock/ManagedConnectionFactoryMock.java
@@ -1,0 +1,67 @@
+/*
+ * Copyright (c) 2025 Eclipse Foundation and/or its affiliates. All rights reserved.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License v. 2.0, which is available at
+ * http://www.eclipse.org/legal/epl-2.0.
+ *
+ * This Source Code may also be made available under the following Secondary
+ * Licenses when the conditions for such availability set forth in the
+ * Eclipse Public License v. 2.0 are satisfied: GNU General Public License,
+ * version 2 with the GNU Classpath Exception, which is available at
+ * https://www.gnu.org/software/classpath/license.html.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+ */
+
+package com.sun.enterprise.resource.pool.mock;
+
+import jakarta.resource.ResourceException;
+import jakarta.resource.spi.ConnectionManager;
+import jakarta.resource.spi.ConnectionRequestInfo;
+import jakarta.resource.spi.ManagedConnection;
+import jakarta.resource.spi.ManagedConnectionFactory;
+
+import java.io.PrintWriter;
+import java.util.Set;
+
+import javax.security.auth.Subject;
+
+/**
+ * Mock class without any implementation
+ */
+public class ManagedConnectionFactoryMock implements ManagedConnectionFactory {
+
+    private static final long serialVersionUID = 1L;
+
+    @Override
+    public Object createConnectionFactory(ConnectionManager cxManager) throws ResourceException {
+        return null;
+    }
+
+    @Override
+    public Object createConnectionFactory() throws ResourceException {
+        return null;
+    }
+
+    @Override
+    public ManagedConnection createManagedConnection(Subject subject, ConnectionRequestInfo cxRequestInfo)
+            throws ResourceException {
+        return null;
+    }
+
+    @Override
+    public ManagedConnection matchManagedConnections(Set connectionSet, Subject subject,
+            ConnectionRequestInfo cxRequestInfo) throws ResourceException {
+        return null;
+    }
+
+    @Override
+    public void setLogWriter(PrintWriter out) throws ResourceException {
+    }
+
+    @Override
+    public PrintWriter getLogWriter() throws ResourceException {
+        return null;
+    }
+}

--- a/appserver/connectors/connectors-runtime/src/test/java/com/sun/enterprise/resource/pool/mock/ManagedConnectionMock.java
+++ b/appserver/connectors/connectors-runtime/src/test/java/com/sun/enterprise/resource/pool/mock/ManagedConnectionMock.java
@@ -1,0 +1,84 @@
+/*
+ * Copyright (c) 2025 Eclipse Foundation and/or its affiliates. All rights reserved.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License v. 2.0, which is available at
+ * http://www.eclipse.org/legal/epl-2.0.
+ *
+ * This Source Code may also be made available under the following Secondary
+ * Licenses when the conditions for such availability set forth in the
+ * Eclipse Public License v. 2.0 are satisfied: GNU General Public License,
+ * version 2 with the GNU Classpath Exception, which is available at
+ * https://www.gnu.org/software/classpath/license.html.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+ */
+
+package com.sun.enterprise.resource.pool.mock;
+
+import jakarta.resource.ResourceException;
+import jakarta.resource.spi.ConnectionEventListener;
+import jakarta.resource.spi.ConnectionRequestInfo;
+import jakarta.resource.spi.LocalTransaction;
+import jakarta.resource.spi.ManagedConnection;
+import jakarta.resource.spi.ManagedConnectionMetaData;
+
+import java.io.PrintWriter;
+
+import javax.security.auth.Subject;
+import javax.transaction.xa.XAResource;
+
+/**
+ * Mock class without any implementation
+ */
+public class ManagedConnectionMock implements ManagedConnection {
+
+    @Override
+    public Object getConnection(Subject subject, ConnectionRequestInfo cxRequestInfo) throws ResourceException {
+        return null;
+    }
+
+    @Override
+    public void destroy() throws ResourceException {
+    }
+
+    @Override
+    public void cleanup() throws ResourceException {
+    }
+
+    @Override
+    public void associateConnection(Object connection) throws ResourceException {
+    }
+
+    @Override
+    public void addConnectionEventListener(ConnectionEventListener listener) {
+    }
+
+    @Override
+    public void removeConnectionEventListener(ConnectionEventListener listener) {
+    }
+
+    @Override
+    public XAResource getXAResource() throws ResourceException {
+        return null;
+    }
+
+    @Override
+    public LocalTransaction getLocalTransaction() throws ResourceException {
+        return null;
+    }
+
+    @Override
+    public ManagedConnectionMetaData getMetaData() throws ResourceException {
+        return null;
+    }
+
+    @Override
+    public void setLogWriter(PrintWriter out) throws ResourceException {
+    }
+
+    @Override
+    public PrintWriter getLogWriter() throws ResourceException {
+        return null;
+    }
+}

--- a/appserver/connectors/connectors-runtime/src/test/java/com/sun/enterprise/resource/pool/mock/TransactionalResourceMock.java
+++ b/appserver/connectors/connectors-runtime/src/test/java/com/sun/enterprise/resource/pool/mock/TransactionalResourceMock.java
@@ -1,0 +1,80 @@
+/*
+ * Copyright (c) 2025 Eclipse Foundation and/or its affiliates. All rights reserved.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License v. 2.0, which is available at
+ * http://www.eclipse.org/legal/epl-2.0.
+ *
+ * This Source Code may also be made available under the following Secondary
+ * Licenses when the conditions for such availability set forth in the
+ * Eclipse Public License v. 2.0 are satisfied: GNU General Public License,
+ * version 2 with the GNU Classpath Exception, which is available at
+ * https://www.gnu.org/software/classpath/license.html.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+ */
+
+package com.sun.enterprise.resource.pool.mock;
+
+import jakarta.transaction.Transaction;
+
+import javax.transaction.xa.XAResource;
+
+/**
+ * Mock class without any implementation
+ */
+public class TransactionalResourceMock implements com.sun.enterprise.transaction.spi.TransactionalResource {
+
+    @Override
+    public void closeUserConnection() throws Exception {
+    }
+
+    @Override
+    public void enlistedInTransaction(Transaction arg0) throws IllegalStateException {
+    }
+
+    @Override
+    public Object getComponentInstance() {
+        return null;
+    }
+
+    @Override
+    public String getName() {
+        return null;
+    }
+
+    @Override
+    public XAResource getXAResource() {
+        return null;
+    }
+
+    @Override
+    public boolean isEnlisted() {
+        return false;
+    }
+
+    @Override
+    public boolean isEnlistmentSuspended() {
+        return false;
+    }
+
+    @Override
+    public boolean isShareable() {
+        return false;
+    }
+
+    @Override
+    public boolean isTransactional() {
+        return false;
+    }
+
+    @Override
+    public void setComponentInstance(Object arg0) {
+    }
+
+    @Override
+    public boolean supportsXA() {
+        return false;
+    }
+
+}

